diff --git a/torch/CMakeLists.txt b/torch/CMakeLists.txt
index 8d761068d1e..2c925270d80 100644
--- a/torch/CMakeLists.txt
+++ b/torch/CMakeLists.txt
@@ -265,7 +265,7 @@ add_custom_command(
     OUTPUT
     "${TORCH_SRC_DIR}/utils/data/datapipes/datapipe.pyi"
     COMMAND
-    ${CMAKE_COMMAND} -E env PYTHONPATH="${TORCH_ROOT}"
+    ${CMAKE_COMMAND} -E env PYTHONPATH="${TORCH_ROOT}:$ENV{PYTHONPATH}"
     "${Python_EXECUTABLE}" ${TORCH_SRC_DIR}/utils/data/datapipes/gen_pyi.py
     DEPENDS
     "${TORCH_SRC_DIR}/utils/data/datapipes/datapipe.pyi.in"
