diff --git a/scripts/spack_test/CMakeLists.txt b/scripts/spack_test/CMakeLists.txt
index 20274432e4b148ab863439329b63735645b1074f..f71c493d86678cff8e81cc93313554b79508085e 100644
--- a/scripts/spack_test/CMakeLists.txt
+++ b/scripts/spack_test/CMakeLists.txt
@@ -1,5 +1,5 @@
 cmake_minimum_required(VERSION 3.16)
-project(SpackTestGen)
+project(SpackTestGen LANGUAGES NONE)
 set(TEST_LIST_DEF ${CMAKE_CURRENT_SOURCE_DIR}/test_list.def)
 file(STRINGS ${TEST_LIST_DEF} TEST_FILES)
