diff --git a/CMakeLists.txt b/CMakeLists.txt
index fa478b5..926b369 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -990,10 +990,11 @@ ENDIF(USE_HDF5)
 FIND_PACKAGE(CURL)
 ADD_DEFINITIONS(-DCURL_STATICLIB=1)
 INCLUDE_DIRECTORIES(${CURL_INCLUDE_DIRS})
-
+set(IMPORT_CURL "")
 # Define a test flag for have curl library
 IF(CURL_LIBRARIES OR CURL_LIBRARY)
   SET(FOUND_CURL TRUE)
+  SET(IMPORT_CURL )
 ELSE()
   SET(FOUND_CURL FALSE)
 ENDIF()
diff --git a/netCDFConfig.cmake.in b/netCDFConfig.cmake.in
index eece09c..66fae64 100644
--- a/netCDFConfig.cmake.in
+++ b/netCDFConfig.cmake.in
@@ -17,7 +17,7 @@ include("${CMAKE_CURRENT_LIST_DIR}/netCDFTargets.cmake")
 @IMPORT_FIND_DEP@
 @IMPORT_MPI@
 @IMPORT_HDF5@
-
+@IMPORT_CURL@
 # Compiling Options
 #
 set(netCDF_C_COMPILER "@CC_VERSION@")
