diff --git a/CMakeLists.txt b/CMakeLists.txt
index 588a2ce4a..e3772a05a 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1555,24 +1555,6 @@ endif()
 # Brute force, grab all of the dlls from the dependency directory,
 # install them in the binary dir. Grab all of the .libs, put them
 # in the libdir.
-##
-if(MSVC)
-  foreach(CPP ${CMAKE_PREFIX_PATH})
-    file(GLOB COPY_FILES ${CPP}/lib/*.lib)
-  endforeach()
-  install(FILES ${COPY_FILES}
-  DESTINATION ${CMAKE_INSTALL_LIBDIR}
-  COMPONENT dependencies)
-
-  foreach(CPP ${CMAKE_PREFIX_PATH})
-    file(GLOB COPY_FILES ${CPP}/bin/*.dll)
-  endforeach()
-  string(REGEX REPLACE "msv[.*].dll" "" COPY_FILES "${COPY_FILES}")
-  install(FILES ${COPY_FILES}
-  DESTINATION ${CMAKE_INSTALL_BINDIR}
-  COMPONENT dependencies)
-
-endif()
 
 # Subdirectory CMakeLists.txt files should specify their own
 # 'install' files.
