diff --git a/packages/stk/stk_mesh/stk_mesh/base/ConstFieldDataBytes.hpp b/packages/stk/stk_mesh/stk_mesh/base/ConstFieldDataBytes.hpp
index 0642a3c3c85..d04e3598cd3 100644
--- a/packages/stk/stk_mesh/stk_mesh/base/ConstFieldDataBytes.hpp
+++ b/packages/stk/stk_mesh/stk_mesh/base/ConstFieldDataBytes.hpp
@@ -123,7 +123,7 @@ public:
   ConstFieldDataBytes();
   ConstFieldDataBytes(EntityRank entityRank, Ordinal fieldOrdinal, const std::string& fieldName,
                       const DataTraits& dataTraits, Layout dataLayout);
-  virtual ~ConstFieldDataBytes() override = default;
+  KOKKOS_FUNCTION virtual ~ConstFieldDataBytes() override {}
 
   // The AMD ROCm compiler has an "undefined hidden symbol" link error when this copy constructor is defined
   // below, along with all of the other functions.  Not sure why this one is special.
diff --git a/packages/stk/stk_mesh/stk_mesh/base/FieldDataBytes.hpp b/packages/stk/stk_mesh/stk_mesh/base/FieldDataBytes.hpp
index fcd9fe03bcc..9112cbe06c4 100644
--- a/packages/stk/stk_mesh/stk_mesh/base/FieldDataBytes.hpp
+++ b/packages/stk/stk_mesh/stk_mesh/base/FieldDataBytes.hpp
@@ -87,7 +87,7 @@ public:
   FieldDataBytes();
   FieldDataBytes(EntityRank entityRank, Ordinal fieldOrdinal, const std::string& fieldName,
                  const DataTraits& dataTraits, Layout dataLayout);
-  virtual ~FieldDataBytes() override = default;
+  KOKKOS_FUNCTION virtual ~FieldDataBytes() override {}
 
   FieldDataBytes(const FieldDataBytes& fieldData) = default;
   FieldDataBytes(FieldDataBytes&&) = default;
