diff --git a/packages/seacas/libraries/ioss/src/CMakeLists.txt b/packages/seacas/libraries/ioss/src/CMakeLists.txt
index 867e1b4379..414b0b61bf 100644
--- a/packages/seacas/libraries/ioss/src/CMakeLists.txt
+++ b/packages/seacas/libraries/ioss/src/CMakeLists.txt
@@ -82,6 +82,10 @@ TRIBITS_ADD_LIBRARY(
   SOURCES ${SOURCES}
   )
 
+if(WIN32)
+  target_compile_definitions(Ioss PUBLIC __IOSS_WINDOWS__)
+endif()
+
 include(GenerateExportHeader)
 generate_export_header(Ioss)
 target_include_directories(Ioss
