diff --git a/qtdeclarative/src/3rdparty/masm/wtf/MathExtras.h b/qtdeclarative/src/3rdparty/masm/wtf/MathExtras.h
index 6af108c886..cc8039ac98 100644
--- a/qtdeclarative/src/3rdparty/masm/wtf/MathExtras.h
+++ b/qtdeclarative/src/3rdparty/masm/wtf/MathExtras.h
@@ -79,15 +79,6 @@ const double piOverFourDouble = M_PI_4;
 const float piOverFourFloat = static_cast<float>(M_PI_4);
 #endif
 
-#if OS(DARWIN)
-
-// Work around a bug in the Mac OS X libc where ceil(-0.1) return +0.
-inline double wtf_ceil(double x) { return copysign(ceil(x), x); }
-
-#define ceil(x) wtf_ceil(x)
-
-#endif
-
 #if OS(SOLARIS) && __cplusplus < 201103L
 
 namespace std {
diff --git a/qtscript/src/3rdparty/javascriptcore/JavaScriptCore/wtf/MathExtras.h b/qtscript/src/3rdparty/javascriptcore/JavaScriptCore/wtf/MathExtras.h
index bea5896..f1cd444 100644
--- a/qtscript/src/3rdparty/javascriptcore/JavaScriptCore/wtf/MathExtras.h
+++ b/qtscript/src/3rdparty/javascriptcore/JavaScriptCore/wtf/MathExtras.h
@@ -67,16 +67,6 @@ const float piOverFourFloat = 0.785398163397448309616f;
 const double piOverFourDouble = M_PI_4;
 const float piOverFourFloat = static_cast<float>(M_PI_4);
 #endif
-
-#if OS(DARWIN)
-
-// Work around a bug in the Mac OS X libc where ceil(-0.1) return +0.
-inline double wtf_ceil(double x) { return copysign(ceil(x), x); }
-
-#define ceil(x) wtf_ceil(x)
-
-#endif
-
 #if OS(SOLARIS) && __cplusplus < 201103L
 
 namespace std {
