diff --git a/CMakeLists.txt b/CMakeLists.txt
index a8f977c75..0e5b81fbd 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -22,7 +22,7 @@ option(build_parse     "Parses source code and dumps the dependencies between th
 option(build_search    "Build external search tools (doxysearch and doxyindexer)" OFF)
 option(build_doc       "Build user manual (HTML and PDF)" OFF)
 option(build_doc_chm   "Build user manual (CHM)" OFF)
-if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
+if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
     option(use_libc++  "Use libc++ as C++ standard library." ON)
 endif()
 option(use_libclang    "Add support for libclang parsing." OFF)
@@ -81,8 +81,18 @@ if (build_wizard)
     endif()
 endif()
 
-# use C++17 standard for compiling
-set(CMAKE_CXX_STANDARD 17)
+# conditionally use C++20 standard for newer Clang versions
+if ((CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang" AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL "17.0") OR
+    (CMAKE_CXX_COMPILER_ID STREQUAL "Clang" AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL "19.0"))
+  # use C++20 standard for compiling with newer Clang
+  set(CMAKE_CXX_STANDARD 20)
+  # needed for JavaCC with C++20
+  set(JAVA_CC_CHAR_TYPE "char8_t")
+else()
+  # use C++17 standard for compiling
+  set(CMAKE_CXX_STANDARD 17)
+  set(JAVA_CC_CHAR_TYPE "unsigned char")
+endif()
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_EXTENSIONS ON)
 
@@ -170,7 +180,7 @@ if (CMAKE_SYSTEM_NAME MATCHES "Windows")
 endif()
 
 # needed for JavaCC
-set(JAVA_CC_EXTRA_FLAGS "-DJAVACC_CHAR_TYPE=\"unsigned char\"")
+set(JAVA_CC_EXTRA_FLAGS "-DJAVACC_CHAR_TYPE=\"${JAVA_CC_CHAR_TYPE}\"")
 set(CMAKE_CXX_FLAGS       "${CMAKE_CXX_FLAGS}       ${JAVA_CC_EXTRA_FLAGS}")
 set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} ${JAVA_CC_EXTRA_FLAGS}")
 
diff --git a/vhdlparser/VhdlParserErrorHandler.hpp b/vhdlparser/VhdlParserErrorHandler.hpp
index a6303dc5c..0d4a4d378 100644
--- a/vhdlparser/VhdlParserErrorHandler.hpp
+++ b/vhdlparser/VhdlParserErrorHandler.hpp
@@ -50,7 +50,7 @@ class VhdlTokenManagerErrorHandler: public TokenManagerErrorHandler
 
     virtual void lexicalError(bool EOFSeen, int /* lexState */, int errorLine, int /* errorColumn */, const JJString& errorAfter, JJChar curChar, VhdlParserTokenManager* /* token_manager */)
     {
-      warn(m_fileName,errorLine,"Lexical error, Encountered: '%c' after: '%s'",curChar, (EOFSeen? "EOF" : (const char*)errorAfter.c_str()));
+      warn(m_fileName,errorLine,"Lexical error, Encountered: '%c' after: '%s'",(char)curChar, (EOFSeen? "EOF" : (const char*)errorAfter.c_str()));
     }
 
     virtual void lexicalError(const JJString& errorMessage, VhdlParserTokenManager* /* token_manager */)
