diff --git a/cmake/project-config.cmake.in b/cmake/project-config.cmake.in
index 23f997ab..f0d46124 100644
--- a/cmake/project-config.cmake.in
+++ b/cmake/project-config.cmake.in
@@ -15,6 +15,12 @@ set (@PROJECT_VARIANT_NAME@_INCLUDE_DIRS "${_ROOT}/@INCLUDEDIR@")
 set (@PROJECT_VARIANT_NAME@_LIBRARY_DIRS "${_ROOT}/@LIBDIR@")
 set (@PROJECT_VARIANT_NAME@_BINARY_DIRS "${_ROOT}/@BINDIR@")
 
+# Establish dependencies
+if (@ENABLE_TIFF@) #ENABLE_TIFF
+  include(CMakeFindDependencyMacro)
+  find_dependency(TIFF)
+endif()
+
 set (@PROJECT_VARIANT_NAME@_LIBRARIES @PROJECT_VARIANT_NAME@::proj)
 # Read in the exported definition of the library
 include ("${_DIR}/@PROJECT_NAME_LOWER@-targets.cmake")
