--- a/spack-src/numpy/_core/src/common/npy_atomic.h	2025-06-24 16:23:34.000000000 +0900
+++ b/spack-src/numpy/_core/src/common/npy_atomic.h	2025-06-24 16:24:23.000000000 +0900
@@ -43,7 +43,7 @@
 npy_atomic_load_uint8(const npy_uint8 *obj) {
 #ifdef STDC_ATOMICS
     _NPY_USING_STD;
-    return (npy_uint8)atomic_load((const _Atomic(uint8_t)*)obj);
+    return (npy_uint8)atomic_load((_Atomic(uint8_t)*)obj);
 #elif defined(MSC_ATOMICS)
 #if defined(_M_X64) || defined(_M_IX86)
     return *(volatile npy_uint8 *)obj;
@@ -59,7 +59,7 @@
 npy_atomic_load_ptr(const void *obj) {
 #ifdef STDC_ATOMICS
     _NPY_USING_STD;
-    return atomic_load((const _Atomic(void *)*)obj);
+    return atomic_load((_Atomic(void *)*)obj);
 #elif defined(MSC_ATOMICS)
 #if SIZEOF_VOID_P == 8
 #if defined(_M_X64) || defined(_M_IX86)
