Submodule external/timemory contains modified content
diff --git a/external/timemory/source/timemory/tpls/cereal/cereal/types/tuple.hpp b/external/timemory/source/timemory/tpls/cereal/cereal/types/tuple.hpp
index 3f02ba23..ec125506 100644
--- a/external/timemory/source/timemory/tpls/cereal/cereal/types/tuple.hpp
+++ b/external/timemory/source/timemory/tpls/cereal/cereal/types/tuple.hpp
@@ -101,7 +101,7 @@ struct serialize
     template <class Archive, class... Types>
     inline static void apply(Archive& ar, std::tuple<Types...>& tuple)
     {
-        serialize<Height - 1>::template apply(ar, tuple);
+        serialize<Height - 1>::template apply<>(ar, tuple);
         ar(TIMEMORY_CEREAL_NVP_(tuple_element_name<Height - 1>::c_str(),
                                 std::get<Height - 1>(tuple)));
     }
@@ -123,7 +123,7 @@ template <class Archive, class... Types>
 inline void
 TIMEMORY_CEREAL_SERIALIZE_FUNCTION_NAME(Archive& ar, std::tuple<Types...>& tuple)
 {
-    tuple_detail::serialize<std::tuple_size<std::tuple<Types...>>::value>::template apply(
+    tuple_detail::serialize<std::tuple_size<std::tuple<Types...>>::value>::template apply<>(
         ar, tuple);
 }
 }  // namespace cereal
