From: Thomas Bouvier <contact@thomas-bouvier.io>
Subject: Fix build with Cython 3.1

--- a/src/pygrib/_pygrib.pyx
+++ b/src/pygrib/_pygrib.pyx
@@ -420,7 +419,7 @@ cdef class open(object):
             grbs = [self.message(n) for n in xrange(beg,end,inc)]
             self.seek(msg) # put iterator back in original position
             return grbs
-        elif type(key) == int or type(key) == long:
+        elif type(key) == int:
             # for an integer, return a single grib message.
             msg = self.tell()
             grb = self.message(key)
@@ -1925,8 +1924,8 @@ Example usage:
             # if there are no matches for this key, just skip it
             if not size:
                 continue
-            if typ == 'l' or (type(v) == int or type(v) == long):
-                longval = long(v)
+            if typ == 'l' or (type(v) == int):
+                longval = int(v)
                 err = grib_index_select_long(self._gi, key, longval)
                 if err:
                     raise RuntimeError(_get_error_message(err))
