diff --git i/prov/cxi/include/cxip.h w/prov/cxi/include/cxip.h
index 52e9136ac..01b8374ee 100644
--- i/prov/cxi/include/cxip.h
+++ w/prov/cxi/include/cxip.h
@@ -181,7 +181,7 @@
 #define CXIP_MINOR_VERSION		1
 #define CXIP_PROV_VERSION		FI_VERSION(CXIP_MAJOR_VERSION, \
 						   CXIP_MINOR_VERSION)
-#define CXIP_FI_VERSION			FI_VERSION(2, 2)
+#define CXIP_FI_VERSION			FI_VERSION(2, 3)
 #define CXIP_WIRE_PROTO_VERSION		1
 
 #define	CXIP_COLL_MAX_CONCUR		8
