diff --git a/Makefile b/Makefile
index 51fe35d..3480939 100644
--- a/Makefile
+++ b/Makefile
@@ -22,7 +22,7 @@ PREFIX="$(MAKEDIR)\install\"
 !ENDIF
 
 !IF "$(BUILD_TYPE)" == "DEBUG"
-BUILD_CFLAGS = /Zi
+BUILD_CFLAGS = /Zi /fsanitize=address
 BUILD_LINK = /DEBUG
 !ENDIF
 
@@ -58,8 +58,8 @@ setup_test: cl.exe
 
 build_and_check_test_sample : setup_test
 	cd tmp\test
-	cl /c /EHsc ..\..\test\calc.cxx /DCALC_EXPORTS
-	cl /c /EHsc ..\..\test\main.cxx
+	cl /c /EHsc ..\..\test\calc.cxx /DCALC_EXPORTS /I ..\..\test\include
+	cl /c /EHsc ..\..\test\main.cxx /I ..\..\test\include
 	link $(LFLAGS) calc.obj /out:calc.dll /DLL
 	link $(LFLAGS) main.obj calc.lib /out:tester.exe
 	tester.exe
@@ -77,15 +77,34 @@ test_wrapper : build_and_check_test_sample
 	rmdir /q /s tmp_bin
 	cd ..
 
-test_relocate: build_and_check_test_sample
+test_relocate_exe: build_and_check_test_sample
 	cd tmp\test
 	-@ if NOT EXIST "relocate.exe" mklink relocate.exe cl.exe
 	move calc.dll ..\calc.dll
-	relocate.exe tester.exe --deploy --full
-	relocate.exe tester.exe --export --full
+	relocate.exe --pe tester.exe --deploy --full
+	relocate.exe --pe tester.exe --export --full
 	tester.exe
+	move ..\calc.dll calc.dll
+	cd ../..
 
-test: test_wrapper test_relocate 
+test_relocate_dll: build_and_check_test_sample
+	cd tmp/test
+	-@ if NOT EXIST "relocate.exe" mklink relocate.exe cl.exe
+	cd ..
+	mkdir tmp_bin
+	mkdir tmp_lib
+	move test\calc.dll tmp_bin\calc.dll
+	move test\calc.lib tmp_lib\calc.lib
+	test\relocate.exe --pe tmp_bin\calc.dll --coff tmp_lib\calc.lib --export
+	cd test
+	del tester.exe
+	link main.obj ..\tmp_lib\calc.lib /out:tester.exe
+	.\tester.exe
+
+test_and_cleanup: test clean-test
+
+
+test: test_wrapper test_relocate_exe test_relocate_dll
 
 
 clean : clean-test clean-cl
diff --git a/src/commandline.cxx b/src/commandline.cxx
index 770d786..9f8aaa0 100644
--- a/src/commandline.cxx
+++ b/src/commandline.cxx
@@ -37,7 +37,7 @@ int redefinedArgCheck(const std::map<std::string, std::string> &args, const char
 int checkArgumentPresence(const std::map<std::string, std::string> &args, const char * val, bool required = true)
 {
     if (args.find(val) == args.end()) {
-        std::cerr << "Warning! Argument (" << val << ") not present";
+        std::cerr << "Warning! Argument (" << val << ") not present\n";
         if (required) {
             return 0;
         }
@@ -58,19 +58,12 @@ std::map<std::string, std::string> ParseRelocate(const char ** args, int argc) {
             }
             opts.insert(std::pair<std::string, std::string>("pe", args[++i]));
         }
-        else if (endswith((std::string)args[i], ".dll")) {
-            if(redefinedArgCheck(opts, "pe", "pe")) {
-                opts.clear();
-                return opts;
-            }
-            opts.insert(std::pair<std::string, std::string>("pe", args[i]));
-        }
-        else if (endswith((std::string)args[i], ".exe")) {
-            if(redefinedArgCheck(opts, "pe", "pe")) {
+        else if (!strcmp(args[i], "--coff")) {
+            if(redefinedArgCheck(opts, "coff", "--coff")) {
                 opts.clear();
                 return opts;
             }
-            opts.insert(std::pair<std::string, std::string>("pe", args[i]));
+            opts.insert(std::pair<std::string, std::string>("coff", args[++i]));
         }
         else if (!strcmp(args[i], "--full")) {
             if(redefinedArgCheck(opts, "full", "--full")) {
@@ -99,15 +92,20 @@ std::map<std::string, std::string> ParseRelocate(const char ** args, int argc) {
             } 
             opts.insert(std::pair<std::string, std::string>("cmd", "deploy"));
         }
+        else if (!strcmp(args[i], "--debug") || !strcmp(args[i], "-d")) {
+            opts.insert(std::pair<std::string, std::string>("debug", "on"));
+        }
+        else if (!strcmp(args[i], "--verify")) {
+            opts.insert(std::pair<std::string, std::string>("verify", "on"));
+        }
+        else if (!strcmp(args[i], "--report")) {
+            opts.insert(std::pair<std::string, std::string>("report", "report"));
+        }
         else {
             // Unknown argument, warn the user it will not be used
-            std::cerr << "Unknown argument :" << args[i] << "will be ignored\n";
+            std::cerr << "Unknown argument: " << args[i] << " will be ignored\n";
         }
     }
-    if(!checkArgumentPresence(opts, "pe")) {
-        opts.clear();
-        return opts;
-    }
     return opts;
 }
 
@@ -165,29 +163,37 @@ bool print_help()
     std::cout << "      In this case, the CLI of this wrapper is identical to cl|ifx|link.\n";
     std::cout << "      See https://learn.microsoft.com/en-us/cpp/build/reference/c-cpp-building-reference\n";
     std::cout << "\n";
-    std::cout << "      cl.exe /c foo.c";
+    std::cout << "      cl.exe /c foo.c\n";
     std::cout << "\n";
     std::cout << "     To preform relocation, invoke the 'relocate' symlink to this file:\n";
     std::cout << "\n";
     std::cout << "      Options:\n";
-    std::cout << "          [--pe] <path to pe file>                     = PE file to be relocated\n";
+    std::cout << "          --pe <path to pe file>                       = PE (dll/exe) file to be relocated\n";
+    std::cout << "          [--coff <path to coff file>]                 = COFF (import library) file to be relocated\n";
+    std::cout << "                                                         If relocating an exe, this is not required.\n";
     std::cout << "          --full                                       = Relocate dynamic references inside\n";
-    std::cout << "                                                          the dll in addition to re-generating\n";
+    std::cout << "                                                          the pe in addition to re-generating\n";
     std::cout << "                                                          the import library\n";
     std::cout << "                                                          Note: this is assumed to be true if\n";
     std::cout << "                                                           relocating an executable.\n";
     std::cout << "                                                          If an executable is relocated, no import\n";
     std::cout << "                                                          library operations are performed.\n";
+    std::cout << "                                                          When relocating a DLL, the import library for\n";
+    std::cout << "                                                          said library is regenerated and the old imp lib\n";
+    std::cout << "                                                          replaced.\n";
     std::cout << "          --export|--deploy                             = Mutually exclusive command modifier.\n";
     std::cout << "                                                           Instructs relocate to either prepare the\n";
     std::cout << "                                                           dynamic library for exporting to build cache\n";
     std::cout << "                                                           or for extraction from bc onto new host system\n";
     std::cout << "          --report                                      = Report information about the parsed PE/Coff files\n";
+    std::cout << "          --debug|-d                                    = Debug relocate run\n";
+    std::cout << "          --verify                                      = Validates that the given file 'pe' is an import library\n";
     std::cout << "\n";
     std::cout << "     To report on PE/COFF files, invoke the 'reporter' symlink to this executable or use the --report flag when invoking 'relocate'";
     std::cout << "\n";
     std::cout << "     Options:\n";
     std::cout << "         <path to file>                                 = Path to any PE or COFF file\n";
+    std::cout << "     To debug any flavor of this wrapper, set the environment variable SPACK_DEBUG_WRAPPER to any value in the wrapper context.\n";
     std::cout << "\n";
     return true;
 }
@@ -204,4 +210,3 @@ bool CheckAndPrintHelp(const char ** arg, int argc)
     return false;
 
 }
-
diff --git a/src/execute.cxx b/src/execute.cxx
index b89fff3..77d49a6 100644
--- a/src/execute.cxx
+++ b/src/execute.cxx
@@ -72,6 +72,7 @@ void ExecuteCommand::SetupExecute()
  */
 int ExecuteCommand::CreateChildPipes()
 {
+    // Create stdout pipes
     SECURITY_ATTRIBUTES saAttr;
     // Set the bInheritHandle flag so pipe handles are inherited.
     saAttr.nLength = sizeof(SECURITY_ATTRIBUTES);
@@ -80,8 +81,21 @@ int ExecuteCommand::CreateChildPipes()
     this->saAttr = saAttr;
     if( !CreatePipe(&this->ChildStdOut_Rd, &this->ChildStdOut_Wd, &saAttr, 0) )
         return 0;
-    if ( !SetHandleInformation(ChildStdOut_Rd, HANDLE_FLAG_INHERIT, 0) )
+    if ( !SetHandleInformation(this->ChildStdOut_Rd, HANDLE_FLAG_INHERIT, 0) )
         return 0;
+
+    // create stderr pipes
+    SECURITY_ATTRIBUTES saAttrErr;
+    // Set the bInheritHandle flag so pipe handles are inherited.
+    saAttrErr.nLength = sizeof(SECURITY_ATTRIBUTES);
+    saAttrErr.bInheritHandle = TRUE;
+    saAttrErr.lpSecurityDescriptor = NULL;
+    this->saAttrErr = saAttrErr;
+    if( !CreatePipe(&this->ChildStdErr_Rd, &this->ChildStdErr_Wd, &saAttrErr, 0) )
+        return 0;
+    if ( !SetHandleInformation(this->ChildStdErr_Rd, HANDLE_FLAG_INHERIT, 0) )
+        return 0;
+
     return 1;
 }
 
@@ -92,6 +106,7 @@ int ExecuteCommand::CreateChildPipes()
 bool ExecuteCommand::ExecuteToolChainChild()
 {
     LPVOID lpMsgBuf;
+    debug("Executing Command: " + this->ComposeCLI());
     const std::wstring c_commandLine = ConvertAnsiToWide(this->ComposeCLI());
     wchar_t * nc_commandLine = _wcsdup(c_commandLine.c_str());
     if(! CreateProcessW(
@@ -118,6 +133,8 @@ bool ExecuteCommand::ExecuteToolChainChild()
             (LPTSTR) &lpMsgBuf,
             0, NULL
         );
+        std::cerr << "Failed to initiate child process from: " << ConvertWideToANSI(nc_commandLine) << " ";
+        std::cerr << "With error: ";
         std::cerr << (char*)lpMsgBuf << "\n";
         free(nc_commandLine);
         this->cpw_initalization_failure = true;
@@ -132,6 +149,38 @@ bool ExecuteCommand::ExecuteToolChainChild()
     return true;
 }
 
+
+/* 
+ * Reads for the member variable holding a pipe to the wrapped processes'
+ * STDERR and writes either to this processes' STDERR or a file, depending on
+ * how the process wrapper is configured
+ */
+int ExecuteCommand::PipeChildToStdErr()
+{
+    DWORD dwRead, dwWritten;
+    CHAR chBuf[BUFSIZE];
+    BOOL bSuccess = TRUE;
+    HANDLE hParentOut;
+    if (this->write_to_file) {
+        hParentOut = this->fileout;
+    }
+    else {
+        hParentOut = GetStdHandle(STD_ERROR_HANDLE);
+    }
+
+    for (;;)
+    {
+        bSuccess = ReadFile( this->ChildStdErr_Rd, chBuf, BUFSIZE, &dwRead, NULL);
+        if( ! bSuccess || dwRead == 0 ) break;
+
+        bSuccess = WriteFile(hParentOut, chBuf,
+                            dwRead, &dwWritten, NULL);
+        if (! bSuccess ) break;
+    }
+    return !bSuccess;
+}
+
+
 /* 
  * Reads for the member variable holding a pipe to the wrapped processes'
  * STDOUT and writes either to this processes' STDOUT or a file, depending on
@@ -230,9 +279,10 @@ bool ExecuteCommand::Execute(const std::string &filename)
                                     NULL
                                     );
     }
-    int ret_code = this->ExecuteToolChainChild();
+    bool ret_code = this->ExecuteToolChainChild();
     if (ret_code) {
         this->child_out_future = std::async(std::launch::async, &ExecuteCommand::PipeChildToStdout, this);
+        this->child_err_future = std::async(std::launch::async, &ExecuteCommand::PipeChildToStdErr, this);
         this->exit_code_future = std::async(std::launch::async, &ExecuteCommand::ReportExitCode, this);
     }
     return ret_code;
@@ -246,5 +296,7 @@ int ExecuteCommand::Join()
 {
     if(!this->child_out_future.get())
         return -999;
+    if(!this->child_err_future.get())
+        return -999;
     return this->exit_code_future.get();
 }
diff --git a/src/execute.h b/src/execute.h
index 8449517..94c5a74 100644
--- a/src/execute.h
+++ b/src/execute.h
@@ -37,6 +37,7 @@ private:
     void SetupExecute();
     bool ExecuteToolChainChild();
     int PipeChildToStdout();
+    int PipeChildToStdErr();
     int CreateChildPipes();
     int CleanupHandles();
     int ReportExitCode();
@@ -44,15 +45,21 @@ private:
     // pipe from child process stdout
     // to parent std out or file
     std::future<int> child_out_future;
+    // Holds the exit code of the pipe
+    // from child to parent stderr
+    std::future<int> child_err_future;
     // Holds the exit code of the 
     // command wrapped by this class
     std::future<int> exit_code_future;
     std::string ComposeCLI();
     HANDLE ChildStdOut_Rd;
     HANDLE ChildStdOut_Wd;
+    HANDLE ChildStdErr_Rd;
+    HANDLE ChildStdErr_Wd;
     PROCESS_INFORMATION procInfo;
     STARTUPINFOW startInfo;
     SECURITY_ATTRIBUTES saAttr;
+    SECURITY_ATTRIBUTES saAttrErr;
     HANDLE fileout = INVALID_HANDLE_VALUE;
     bool write_to_file;
     bool cpw_initalization_failure = false;
diff --git a/src/ld.cxx b/src/ld.cxx
index c4eb315..7cbe9fb 100644
--- a/src/ld.cxx
+++ b/src/ld.cxx
@@ -33,13 +33,16 @@ int LdInvocation::InvokeToolchain()
     // We're creating a dll, we need to create an appropriate import lib
     if(!link_run.IsExeLink()) {
         std::string basename = link_run.get_name();
-        std::string imp_lib_name = strip(basename, ".dll") + ".lib";
+        std::string imp_lib_name = link_run.get_implib_name();
         std::string dll_name = link_run.get_mangled_out();
-        std::string abs_out_imp_lib_name = basename + ".dll-abs.lib";
+        std::string abs_out_imp_lib_name = imp_lib_name + ".dll-abs.lib";
+        std::string def_file = link_run.get_def_file();
+        std::string def_line = "-def";
+        def_line += !def_file.empty() ? ":" + def_file : "";
         // create command line to generate new import lib
         this->rpath_executor = ExecuteCommand("lib.exe", this->ComposeCommandLists(
             {
-                {"-def", "-name:" + dll_name, "-out:"+ abs_out_imp_lib_name},
+                {def_line, "-name:" + dll_name, "-out:"+ abs_out_imp_lib_name},
                 this->obj_args,
                 this->lib_args,
                 this->lib_dir_args,
@@ -52,13 +55,25 @@ int LdInvocation::InvokeToolchain()
         }
         CoffReaderWriter cr(abs_out_imp_lib_name);
         CoffParser coff(&cr);
-        coff.Parse();
+        if(!coff.Parse()) {
+            debug("Failed to parse COFF file: " + abs_out_imp_lib_name);
+            return -9;
+        }
         if(!coff.NormalizeName(dll_name)){
+            debug("Failed to normalize name for COFF file: " + abs_out_imp_lib_name);
             return -9;
         }
-        std::remove(imp_lib_name.c_str());
-        std::rename(abs_out_imp_lib_name.c_str(), imp_lib_name.c_str());
-
+        debug("Renaming library from " + abs_out_imp_lib_name + " to " + imp_lib_name);
+        int remove_exitcode = std::remove(imp_lib_name.c_str());
+        if(remove_exitcode) {
+            debug("Failed to remove original import library with exit code: " + remove_exitcode);
+            return -10;
+        }
+        int rename_exitcode = std::rename(abs_out_imp_lib_name.c_str(), imp_lib_name.c_str());
+        if(rename_exitcode) {
+            debug("Failed to rename temporary import library with exit code: " + rename_exitcode);
+            return -11;
+        }
     }
     return ret_code;
 }
diff --git a/src/main.cxx b/src/main.cxx
index a1a6617..829847f 100644
--- a/src/main.cxx
+++ b/src/main.cxx
@@ -25,16 +25,61 @@ int main(int argc, const char* argv[]) {
         bool deploy = !(patch_args.find("cmd") == patch_args.end()) 
             && patch_args.at("cmd") == "deploy";
         bool report = !(patch_args.find("report") == patch_args.end());
-        bool is_exe = endswith(patch_args.at("pe"), ".exe");
+        bool has_pe = !(patch_args.find("pe") == patch_args.end());
+        bool is_exe = has_pe ? endswith(patch_args.at("pe"), ".exe") : false;
+        bool debug = !(patch_args.find("debug") == patch_args.end());
+        bool is_validate = !(patch_args.find("verify") == patch_args.end());
+        bool has_coff = !(patch_args.find("coff") == patch_args.end());
         // Without full with a DLL we re-produce the import lib from the
         // relocated DLL, but with an EXE there is nothing to do
+        if(!has_coff && !has_pe) {
+            std::cout << "No binaries to operate on... nothing to do\n";
+            return -1;
+        }
         if (is_exe && !full)
         {
-            std::cerr << "Executable file provided but --full not specified, nothing to do...\n";
+            std::cout << "Executable file provided but --full not specified, nothing to do...\n";
+            return -1;
+        }
+        // The only scenario its ok to have a dll/exe and no coff is when we're creating a cache
+        // entry
+        if (!is_exe && !has_coff && !deploy) {
+            std::cout << "Attempting to relocate DLL without coff file, please provide a coff file.\n";
+            return -1;
+        }
+        if (is_validate && !has_coff) {
+            std::cout << "Attempting to validate without a coff file, nothing to validate\n";
+            return -1;
+        }
+        if (report && !has_coff) {
+            std::cout << "Attempting to report without a binary, nothing to report...\n";
+            return -1;
+        }
+        if (!(is_validate || report) && !has_pe) {
+            std:: cout << "Attempting to perform relocation without a PE file, please provide one.\n";
+            return -1;
+        }
+        if (is_validate) {
+            return CoffParser::Validate(patch_args.at("coff"));
+        }
+        if (report) {
+            CoffReaderWriter cr(patch_args.at("coff"));
+            CoffParser coff(&cr);
+            if(!reportCoff(coff)) {
+                return 1;
+            }
             return 0;
         }
-        LibRename rpath_lib(patch_args.at("pe"), full, deploy, true, report);
-        if(!rpath_lib.ExecuteRename()){
+        DEBUG = debug;
+        std::unique_ptr<LibRename> rpath_lib;
+        if (has_coff)
+        {
+            rpath_lib = std::make_unique<LibRename>(patch_args.at("pe"), patch_args.at("coff"), full, deploy, true);
+        }
+        else {
+            rpath_lib = std::make_unique<LibRename>(patch_args.at("pe"), full, deploy, true);
+        }
+        if(!rpath_lib->ExecuteRename()){
             std::cerr << "Library rename failed\n";
             return 9;
         }
diff --git a/src/toolchain.cxx b/src/toolchain.cxx
index b71a167..b4fb3a8 100644
--- a/src/toolchain.cxx
+++ b/src/toolchain.cxx
@@ -6,6 +6,8 @@
 #include "toolchain.h"
 
 #include <sstream>
+#include <typeinfo>
+
 
 ToolChainInvocation::ToolChainInvocation(std::string command, char const* const* cli) :
     command(command)
@@ -42,10 +44,12 @@ int ToolChainInvocation::InvokeToolchain() {
     this->executor = ExecuteCommand(  this->command,
                                       commandLine
                                     );
+    debug("Setting up executor for " + std::string(typeid(*this).name()) + "toolchain");
+    debug("Toolchain: " + this->command);
     // Run first pass of command as requested by caller
     int ret_code = this->executor.Execute();
     if(!ret_code) {
-        std::cerr << "Unable to launch process \n";
+        std::cerr << "Unable to launch toolchain process \n";
         return -9999;
     }
     return this->executor.Join();
@@ -56,7 +60,7 @@ void ToolChainInvocation::ParseCommandArgs(char const* const* cli) {
     // Includes come first
     for( char const* const* c = cli; *c; c++ ){
         std::string arg = std::string(*c);
-        if ( startswith(arg, "/I") ) {
+        if ( startswith(arg, "/I") || startswith(arg, "-I") ) {
             // We have an include arg
             // can have an optional space
             // check if there are characters after
@@ -64,10 +68,12 @@ void ToolChainInvocation::ParseCommandArgs(char const* const* cli) {
             // argument to be the include
             if (arg.size() > 2)
                 this->include_args.push_back(arg);
-            else
+            else {
+                this->include_args.push_back(arg);
                 this->include_args.push_back(std::string(*(++c)));
+            }
         }
-        else if( endswith(arg, ".lib") )
+        else if( endswith(arg, ".lib") && (arg.find("implib:") == std::string::npos))
             // Lib args are just libraries
             // provided like system32.lib on the
             // command line.
@@ -102,6 +108,8 @@ StrList ToolChainInvocation::ComposeCommandLists(std::vector<StrList> command_ar
     StrList commandLine;
     for(auto arg_list : command_args)
     {
+        // Ensure arguments are appropriately quoted
+        quoteList(arg_list);
         commandLine.insert(commandLine.end(), arg_list.begin(), arg_list.end());
     }
     return commandLine;
diff --git a/src/utils.cxx b/src/utils.cxx
index d96876e..3300102 100644
--- a/src/utils.cxx
+++ b/src/utils.cxx
@@ -162,8 +162,7 @@ void StripPathAndExe(std::string &command) {
 
 void StripExe(std::string &command) {
     // Normalize command to lowercase to avoid parsing issues
-    std::transform(command.begin(), command.end(), command.begin(),
-        [](unsigned char c){ return std::tolower(c); });
+    lower(command);
     std::string::size_type loc = command.rfind(".exe");
     if ( std::string::npos != loc && loc + 4 == command.length() )
         command.erase(loc);
@@ -173,6 +172,29 @@ void StripPath(std::string &command) {
     command.erase(0, command.find_last_of("\\/") + 1);
 }
 
+/**
+ * Converts a string to lowercase
+ * 
+ * \arg str - string to be made lowercase
+ */
+void lower(std::string &str) {
+    std::transform(str.begin(), str.end(), str.begin(),
+        [](unsigned char c){ return std::tolower(c); });
+}
+
+
+std::string quoteAsNeeded(std::string &str) {
+    if (str.find_first_of(" &<>|()") != std::string::npos) {
+        // There are spaces or special characters in string, quote it
+        return "\"" + str + "\"";
+    }
+    return str;
+}
+
+
+void quoteList(StrList &args) {
+    std::transform(args.begin(), args.end(), args.begin(), quoteAsNeeded);
+}
 
 /**
  * Given an environment variable name
@@ -235,7 +257,7 @@ std::string basename(const std::string &file)
 {
     std:size_t last_path = file.find_last_of("\\")+1;
     if (last_path == std::string::npos) {
-        return std::string();
+        return file;
     }
     return file.substr(last_path);
 }
@@ -276,6 +298,16 @@ DWORD RvaToFileOffset(PIMAGE_SECTION_HEADER &section_header, DWORD number_of_sec
 }
 
 
+void debug(std::string dbgStmt) {
+    if (DEBUG || getenv("SPACK_DEBUG_WRAPPER")) {
+        std::cout << "DEBUG: " << dbgStmt << "\n";
+    }
+}
+
+void debug(char * dbgStmt, int len) {
+    debug(std::string(dbgStmt, len));
+}
+
 std::string reportLastError()
 {
     DWORD error = GetLastError();
@@ -412,6 +444,19 @@ DWORD ToLittleEndian(DWORD val)
     return little_endian_val;
 }
 
+int get_slash_name_length(char *slash_name)
+{
+    if(slash_name == nullptr) {
+        return 0;
+    }
+    int len = 0;
+    // Maximum length for a given name in the PE/COFF format is 143 chars
+    while(slash_name[len] != '/' && len < 144) {
+        ++len;
+    }
+    return len;
+}
+
 char * findstr(char *search_str, const char * substr, int size)
 {
     char * search = search_str;
diff --git a/src/utils.h b/src/utils.h
index 6dc2222..a5376bd 100644
--- a/src/utils.h
+++ b/src/utils.h
@@ -59,6 +59,9 @@ StrList split(const std::string &s, const std::string &delim);
 //Strips substr off the RHS of the larger string
 std::string strip(const std::string& s, const std::string &substr);
 
+//Strips substr of LHS of the larger string
+std::string lstrip(const std::string& s, const std::string &substr);
+
 // Joins vector of strings by join character
 std::string join(const StrList &strs, const std::string &join_char = " ");
 
@@ -78,10 +81,22 @@ void StripExe(std::string &command);
 // resulting in a parentless, non exe extensioned path
 void StripPathAndExe(std::string &command);
 
+// Make str lowercase
+void lower(std::string &str);
+
+// Given a string containing something terminated by a 
+// forward slash, get the length of the substr terminated
+// by /
+int get_slash_name_length(char *slash_name);
+
 // Implementation of strstr but serch is bounded at size and
 // does not terminate on the first read nullptr
 char * findstr(char * search_str, const char * substr, int size);
 
+// Adds quote to relevent strings in a list of strings
+// Strings to be quoted contain: spaces, or any of &<>|()
+void quoteList(StrList &args);
+
 // FS/Path helpers //
 
 // Returns current working directory
@@ -109,6 +124,12 @@ std::string reportLastError();
 // files in big endian format
 DWORD ToLittleEndian(DWORD val);
 
+// Operating Utils //
+
+void debug(std::string dbgStmt);
+
+void debug(char * dbgStmt, int len);
+
 /**
  * Library Searching utility class
  *  Collection of heuristics and logic surrounding library
@@ -132,3 +153,5 @@ public:
     std::string FindLibrary(const std::string &lib_name, const std::string &lib_path);
     void EvalSearchPaths();
 };
+
+static bool DEBUG = false;
diff --git a/src/winrpath.cxx b/src/winrpath.cxx
index 6eb7863..b2cad80 100644
--- a/src/winrpath.cxx
+++ b/src/winrpath.cxx
@@ -160,24 +160,45 @@ LinkerInvocation::LinkerInvocation(const StrList &linkLine)
 void LinkerInvocation::Parse()
 {
     for (auto token = this->tokens.begin(); token != this->tokens.end(); ++token) {
-        if (endswith(*token, ".lib")) {
+        std::string normalToken = *token;
+        lower(normalToken);
+        // implib specifies the eventuall import libraries name
+        // and thus will contain a ".lib" extension, which
+        // the next check will process as a library argument
+        if (normalToken.find("implib:") != std::string::npos) {
+            // If there was nothing after the ":", the
+            // previous link command would have failed
+            // and : is not a legal character in a name
+            // guarantees this split command produces a vec of
+            // len 2
+            StrList implibLine = split(*token, ":");
+            this->implibname = implibLine[1];
+        }
+        else if (endswith(normalToken, ".lib")) {
             this->libs.push_back(*token);
         }
-        else if (*token == "/dll" || *token == "/DLL") {
+        else if (normalToken == "/dll" || normalToken == "-dll") {
             this->is_exe = false;
         }
-        else if (startswith(*token, "-out") || startswith(*token, "/out")) {
+        else if (startswith(normalToken, "-out") || startswith(normalToken, "/out")) {
             this->output = split(*token, ":")[1];
         }
-        else if (endswith(*token, ".obj")) {
+        else if (endswith(normalToken, ".obj")) {
             this->objs.push_back(*token);
         }
+        else if (normalToken.find("def:") != std::string::npos) {
+            StrList defLine = split(*token, ":");
+            this->def_file = defLine[1];
+        }
     }
     std::string ext = this->is_exe ? ".exe" : ".dll";
     if (this->output.empty()){
         this->output = strip(this->objs.front(), ".obj") + ext;
     }
     this->name = strip(this->output, ext);
+    if (this->implibname.empty()) {
+        this->implibname = this->name + ".lib";
+    }
 }
 
 std::string LinkerInvocation::get_name()
@@ -185,6 +206,16 @@ std::string LinkerInvocation::get_name()
     return this->name;
 }
 
+std::string LinkerInvocation::get_implib_name()
+{
+    return this->implibname;
+}
+
+std::string LinkerInvocation::get_def_file()
+{
+    return this->def_file;
+}
+
 std::string LinkerInvocation::get_out()
 {
     return this->output;
@@ -330,7 +361,10 @@ bool CoffParser::Parse()
         std::streampos offset = this->coffStream->tell();
         this->coffStream->ReadHeader(header);
         this->coffStream->ReadMember(header, member);
-        this->ParseData(header, member);
+        if (!this->ParseData(header, member)) {
+            this->verified = true;
+            return false;
+        }
         coff_entry entry;
         entry.header = header;
         entry.member = member;
@@ -347,6 +381,22 @@ bool CoffParser::Parse()
     return true;
 }
 
+int CoffParser::Verify()
+{
+    bool parseStatus = this->Parse();
+    if(!parseStatus && !this->verified) {
+        // actual error in parsing the library
+        return 2;
+    }
+    else if(!parseStatus && this->verified) {
+        // library is valid, it's just a static
+        // lib, not an import
+        return 1;
+    }
+    // otherwise, successful, it's an import lib
+    return 0;
+}
+
 /**
  * Parses a member section in the form of a short format import section
  *  based on the COFF structure scheme
@@ -468,6 +518,17 @@ void CoffParser::ParseSecondLinkerMember(coff_member *member)
     member->second_link = sl;
 }
 
+
+namespace {
+    bool nameCheck(BYTE* name)
+    {
+        int nameLen = get_slash_name_length((char*)name);
+        if(findstr((char*)name, ".obj", nameLen)) {
+            return false;
+        }
+        return true;
+    }
+}
 /**
  * Drive the parsing of the "data" section of an import library member
  * 
@@ -482,7 +543,7 @@ void CoffParser::ParseSecondLinkerMember(coff_member *member)
  * \param header A pointer to the archive member header corresponding to the member being parsed
  * \param member A pointer to the member data being parsed
  */
-void CoffParser::ParseData(PIMAGE_ARCHIVE_MEMBER_HEADER header, coff_member *member)
+bool CoffParser::ParseData(PIMAGE_ARCHIVE_MEMBER_HEADER header, coff_member *member)
 {
     IMPORT_OBJECT_HEADER * p_imp_header = (IMPORT_OBJECT_HEADER *)member->data;
     if((p_imp_header->Sig1 == IMAGE_FILE_MACHINE_UNKNOWN) && (p_imp_header->Sig2 == IMPORT_OBJECT_HDR_SIG2)) {
@@ -490,6 +551,9 @@ void CoffParser::ParseData(PIMAGE_ARCHIVE_MEMBER_HEADER header, coff_member *mem
         this->ParseShortImport(member);
     }
     else if (!strncmp((char*)header->Name, IMAGE_ARCHIVE_LINKER_MEMBER, 16)) {
+        if(!nameCheck(header->Name)){
+            return false;
+        }
         if (!this->coff.read_first_linker) {
             this->ParseFirstLinkerMember(member);
             this->coff.read_first_linker = true;
@@ -499,13 +563,36 @@ void CoffParser::ParseData(PIMAGE_ARCHIVE_MEMBER_HEADER header, coff_member *mem
         }
     }
     else if (!strncmp((char*)header->Name, IMAGE_ARCHIVE_LONGNAMES_MEMBER, 16)) {
-        // Long names member doesn't provide us anything useful to parse
-        // at this stage
-        return;
+        // Check the long names member for values, if so, check the extension has a dll
+        if (!this->ValidateLongName(member, atoi((char*)header->Size))) {
+            return false;
+        }
+        member->is_longname = true;
     }
     else {
+        if(!nameCheck(header->Name)) {
+            return false;
+        }
         this->ParseFullImport(member);
     }
+    return true;
+}
+
+bool CoffParser::ValidateLongName(coff_member* member, int size)
+{
+    if (!member->data) {
+        // If we have no member, by virtue of correctly processing
+        // the header to get to this point
+        // we have a valid header
+        return true;
+    }
+    // If a name has an object file, this is not an import
+    // member
+    char * objRes = findstr(member->data, ".obj", size);
+    if (!objRes) {
+        return true;
+    }
+    return false;
 }
 
 
@@ -553,11 +640,7 @@ void CoffParser::NormalizeSectionNames(const std::string &name, char* section, c
             char * new_name = new char[name_len];
             strncpy(new_name, section_search_start, name_len);
             replace_special_characters(new_name, name_len);
-            this->coffStream->seek(0);
-            this->coffStream->seek(section_data_start_offset + offset);
-            // Reduce name len by one to prevent writing out the null terminator
-            // that is part of the new_name
-            this->coffStream->write(new_name, name_len);    
+            this->writeRename(new_name, name_len, section_data_start_offset + offset);
             delete new_name;
             section_search_start += name_len+1;
             offset = section_search_start - section;                
@@ -565,6 +648,18 @@ void CoffParser::NormalizeSectionNames(const std::string &name, char* section, c
     }
 }
 
+void CoffParser::writeRename(char* name, const int size, const int loc)
+{
+    this->coffStream->seek(0);
+    this->coffStream->seek(loc);
+    this->coffStream->write(name, size);
+}
+
+bool CoffParser::validateName(char* old_name, std::string new_name)
+{
+    return !strcmp(old_name, new_name.c_str());
+}
+
 /**
  * Normalizes mangled DLL names that represent absolute paths in COFF
  * binary files
@@ -609,18 +704,12 @@ bool CoffParser::NormalizeName(std::string &name)
             // We know it exists at this point due to the success of the conditional above
             char* long_name = new char[long_name_len+1];
             strncpy(long_name, this->coff.members[2].member->data+longname_offset, long_name_len+1);
-            // Ensure Dll name is the one we're looking to perform the rename for
-            if (!strcmp(name.c_str(), long_name) && !long_name_renamed) {
+            if (this->validateName(long_name, name) && !long_name_renamed) {
                 // If so, unmangle it
                 replace_special_characters(long_name, long_name_len+1);
                 // offset of actual longname member
                 int offset = std::streamoff(this->coff.members[2].offset);
-                this->coffStream->seek(0);
-                // Seek to longname header
-                this->coffStream->seek(offset);
-                // Seek to offset within longname member for a given import name
-                this->coffStream->seek(sizeof(IMAGE_ARCHIVE_MEMBER_HEADER) + longname_offset, std::ios_base::cur);
-                this->coffStream->write(long_name, long_name_len+1);
+                this->writeRename(long_name, long_name_len+1, offset + sizeof(IMAGE_ARCHIVE_MEMBER_HEADER) + longname_offset);
                 long_name_renamed = true;
             }
             delete long_name;
@@ -636,7 +725,7 @@ bool CoffParser::NormalizeName(std::string &name)
                 strcpy(new_name, mem.member->short_member->short_dll);
                 replace_special_characters(new_name, name_len);
                 // ensure it's the name we're looking to rename
-                if(!strcmp(name.c_str(), mem.member->short_member->short_dll)) {
+                if(this->validateName(mem.member->short_member->short_dll, name)) {
                     // Member offset in file
                     int offset = std::streamoff(mem.offset);
                     // Member header offset
@@ -647,9 +736,7 @@ bool CoffParser::NormalizeName(std::string &name)
                     // Next is the symbol name, which is a null terminated string
                     // +1 to preserve the null terminator in the coff member
                     offset += strlen(mem.member->short_member->short_name) + 1;
-                    this->coffStream->seek(0);
-                    this->coffStream->seek(offset);
-                    this->coffStream->write(new_name, strlen(new_name));
+                    this->writeRename(new_name, strlen(new_name), offset);
                 }
                 delete new_name;
             }
@@ -689,9 +776,7 @@ bool CoffParser::NormalizeName(std::string &name)
                             char * new_no_ext_name = new char[name_len];
                             strncpy(new_no_ext_name, string_table_start, name_len);
                             replace_special_characters(new_no_ext_name, name_len);
-                            this->coffStream->seek(0);
-                            this->coffStream->seek(relative_string_table_start_offset + offset);
-                            this->coffStream->write(new_no_ext_name, name_len);
+                            this->writeRename(new_no_ext_name, name_len, relative_string_table_start_offset + offset);
                             delete new_no_ext_name;
                         }
                     }
@@ -750,19 +835,27 @@ void CoffParser::ReportShortImportMember(short_import_member *si)
 }
 
 
+void CoffParser::ReportLongName(char * data)
+{
+    std::cout << "DLL: " << data << "\n";
+}
+
 void CoffParser::Report()
 {
     for (auto mem: this->coff.members) {
-        reportArchiveHeader(mem.header);
-        if(mem.member->long_member) {
-            this->ReportLongImportMember(mem.member->long_member);
-        }
-        else if(mem.member->short_member) {
-            this->ReportShortImportMember(mem.member->short_member);
+        if(mem.member->is_longname) {
+            this->ReportLongName(mem.member->data);
         }
     }
 }
 
+int CoffParser::Validate(std::string &coff)
+{
+    CoffReaderWriter cr(coff);
+    CoffParser coffp(&cr);
+    return coffp.Verify();
+}
+
 /**
  * Reports information about parsed coff file
  * 
@@ -978,12 +1071,17 @@ bool LibRename::FindDllAndRename(HANDLE &pe_in)
  * \param replace a flag indicating if we're replacing the renamed import lib or making a copy with absolute dll names
  * \param report a flag indicating if we should be reporting the contents of the PE/COFF file we're parsing to stdout
 */
-LibRename::LibRename(std::string pe, bool full, bool deploy, bool replace, bool report)
+LibRename::LibRename(std::string pe, bool full, bool deploy, bool replace)
 : replace(replace), full(full), pe(pe), deploy(deploy)
+{}
+
+LibRename::LibRename(std::string pe, std::string coff, bool full, bool deploy, bool replace)
+: replace(replace), full(full), pe(pe), deploy(deploy), coff(coff)
 {
-    this->name = stem(this->pe);
     this->is_exe = endswith(this->pe, ".exe");
-    this->def_file = this->name + ".def";
+    std::string coff_path = stem(this->coff);
+    this->tmp_def_file = coff_path + "-tmp.def";
+    this->def_file = coff_path + ".def";
     this->def_executor = ExecuteCommand("dumpbin.exe", {this->ComputeDefLine()});
     this->lib_executor = ExecuteCommand("lib.exe", {this->ComputeRenameLink()});
 }
@@ -992,11 +1090,11 @@ LibRename::LibRename(std::string pe, bool full, bool deploy, bool replace, bool
  * Creates the line to be provided to dumpbin.exe to produce the exports of a given
  * dll in the case where we do not have access to the original link line
  * 
- * Produces something like `/EXPORTS <name of pe file`
+ * Produces something like `/EXPORTS <name of coff file>`
  */
 std::string LibRename::ComputeDefLine()
 {
-    return "/EXPORTS " + this->pe;
+    return "/NOLOGO /EXPORTS " + this->coff;
 }
 
 /**
@@ -1005,9 +1103,51 @@ std::string LibRename::ComputeDefLine()
  * 
  * Returns the return code of the Def file computation operation
  */
-int LibRename::ComputeDefFile()
+bool LibRename::ComputeDefFile()
 {
-    return this->def_executor.Execute(this->def_file);
+    this->def_executor.Execute(this->tmp_def_file);
+    int res = this->def_executor.Join();
+    if(res) {
+        return false;
+    }
+    // Need to process the produced def file because it's wrong
+    // Open input file
+    std::ifstream inputFile(this->tmp_def_file);
+    if (!inputFile.is_open()) {
+        std::cerr << "Error: Could not open input file " << tmp_def_file << std::endl;
+        return false;
+    }
+
+    // Open output file
+    std::ofstream outputFile(this->def_file);
+    if (!outputFile.is_open()) {
+        std::cerr << "Error: Could not open output file " << this->def_file << std::endl;
+        return false;
+    }
+
+    // Write the standard .def file header
+    // You might want to get the DLL name dynamically from the input filename or dumpbin output
+    outputFile << "EXPORTS\n";
+
+    std::string line;
+    // First 8 lines are templated output
+    // skip them
+    for (int i = 0; i < 8; ++i) { // Adjust this number if the header changes across dumpbin versions
+        if (!std::getline(inputFile, line)) break;
+    }
+    while (std::getline(inputFile, line)) {
+        if (line.empty()) {
+            continue;
+        } 
+        else if(line.find("Summary") != std::string::npos) { // Skip header in export block if still present
+            break;
+        }
+        outputFile << "    " << lstrip(line, "            ") << std::endl;
+    }
+    inputFile.close();
+    outputFile.close();
+    std::remove(this->tmp_def_file.c_str());
+    return true;
 }
 
 /**
@@ -1037,19 +1177,23 @@ bool LibRename::ExecuteRename()
     // recompute the .def and .lib for dlls
     // exes do not typically have import libs so we don't handle
     // that case
-    if(!this->deploy && !this->is_exe){
+    // We do not bother with defs for things that don't have
+    // import libraries
+    if(!this->deploy && !this->coff.empty()){
         // Extract DLL 
-        if(this->ComputeDefFile()) {
+        if(!this->ComputeDefFile()) {
+            debug("Failed to compute def file");
             return false;
         }
         if(!this->ExecuteLibRename()) {
+            debug("Failed to create and rename import lib");
             return false;
         }
     }
     if (this->full) {
         if(!this->ExecutePERename()) {
             std::cerr << "Unable to execute rename of "
-                "referenced components in PE file: " << this->name << "\n";
+                "referenced components in PE file: " << this->pe << "\n";
             return false;
         }
     }
@@ -1069,20 +1213,24 @@ bool LibRename::ExecuteLibRename()
     this->lib_executor.Execute();
     int ret_code = this->lib_executor.Join();
     if(ret_code != 0) {
-        std::cerr << "Lib Rename failed" << reportLastError() << "\n";
+        std::cerr << "Lib Rename failed with exit code: " << ret_code << "\n";
         return false;
     }
+    // replace former .lib with renamed .lib
+    std::remove(this->coff.c_str());
+    std::rename(this->new_lib.c_str(), this->coff.c_str());
     // import library has been generated with
     // mangled abs path to dll -
     // unmangle it
-    CoffReaderWriter cr(this->new_lib);
+    CoffReaderWriter cr(this->coff);
     CoffParser coff(&cr);
     if (!coff.Parse()) {
         std::cerr << "Unable to parse generated import library {" << this->new_lib <<"}\n";
         return false;
     }
-    if(!coff.NormalizeName(mangle_name(this->pe) )) {
-        std::cerr << "Unable to normalize name\n";
+    std::string mangledName = mangle_name(this->pe);
+    if(!coff.NormalizeName(mangledName)) {
+        std::cerr << "Unable to normalize name: " << mangledName << "\n";
         return false;
     }
     return true;
@@ -1094,10 +1242,10 @@ bool LibRename::ExecuteLibRename()
  */
 bool LibRename::ExecutePERename()
 {
-    LPCWSTR lib_name = ConvertAnsiToWide(this->pe).c_str();
-    HANDLE pe_handle = CreateFileW(lib_name, (GENERIC_READ|GENERIC_WRITE), FILE_SHARE_READ, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+    std::wstring pe_path = ConvertAnsiToWide(this->pe);
+    HANDLE pe_handle = CreateFileW(pe_path.c_str(), (GENERIC_READ|GENERIC_WRITE), FILE_SHARE_READ, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
     if (!pe_handle || pe_handle == INVALID_HANDLE_VALUE){
-        std::cerr << "Unable to acquire file handle to "<< lib_name << ": " << reportLastError() << "\n";
+        std::cerr << "Unable to acquire file handle to "<< pe_path.c_str() << ": " << reportLastError() << "\n";
         return false;
     }
     return this->FindDllAndRename(pe_handle);
@@ -1131,13 +1279,14 @@ std::string LibRename::ComputeRenameLink()
     line += this->def_file + " ";
     line += "-name:";
     line += mangle_name(this->pe) + " ";
-    std::string name(stem(this->pe));
+    std::string name(stem(this->coff));
     if (!this->replace){
         this->new_lib = name + ".abs-name.lib";
     }
     else {
-        this->new_lib = this->pe;
+        // Name must be different
+        this->new_lib = name+"-tmp.lib";
     }
-    line += "-out:\""+ this->new_lib + "\"" + " " + this->pe;
+    line += "-out:\""+ this->new_lib + "\"";
     return line;
 }
diff --git a/src/winrpath.h b/src/winrpath.h
index d2657b5..e440982 100644
--- a/src/winrpath.h
+++ b/src/winrpath.h
@@ -92,8 +92,10 @@ typedef struct coff_member {
     first_linker_member *first_link;
     second_linker_member *second_link;
     bool is_short;
+    bool is_longname;
     coff_member() {
         this->is_short = false;
+        this->is_longname = false;
         this->first_link = NULL;
         this->second_link = NULL;
         this->short_member = NULL;
@@ -177,22 +179,28 @@ class CoffParser {
 private:
     CoffReaderWriter* coffStream;
     coff coff;
-    void ParseData(PIMAGE_ARCHIVE_MEMBER_HEADER header, coff_member *member);
+    bool verified = false;
+    bool ParseData(PIMAGE_ARCHIVE_MEMBER_HEADER header, coff_member *member);
     void ParseShortImport(coff_member *member);
     void ParseFullImport(coff_member *member);
     void ParseFirstLinkerMember(coff_member *member);
     void ParseSecondLinkerMember(coff_member *member);
     void ReportLongImportMember(long_import_member *li);
     void ReportShortImportMember(short_import_member *si);
+    void ReportLongName(char * data);
     void NormalizeLinkerMember(const std::string &name, const int &base_offset, const int &offset, const char * strings, const DWORD symbols);
     void NormalizeSectionNames(const std::string &name, char* section, const DWORD &section_data_start_offset, int data_size);
-
+    bool ValidateLongName(coff_member *member, int size);
+    void writeRename(char *name, const int size, const int loc);
+    bool validateName(char *old_name, std::string new_name);
 public:
     CoffParser(CoffReaderWriter * cr);
     ~CoffParser() = default;
     bool Parse();
     bool NormalizeName(std::string &name);
     void Report();
+    int Verify();
+    static int Validate(std::string &coff);
 };
 
 class LinkerInvocation {
@@ -205,10 +213,14 @@ public:
     std::string get_name();
     std::string get_out();
     std::string get_mangled_out();
+    std::string get_implib_name();
+    std::string get_def_file();
 private:
     std::string line;
     StrList tokens;
     std::string name;
+    std::string implibname;
+    std::string def_file;
     std::string output;
     StrList libs;
     StrList objs;
@@ -218,11 +230,12 @@ private:
 
 class LibRename {
 public:
-    LibRename(std::string pe, bool full, bool deploy, bool replace, bool report);
+    LibRename(std::string pe, std::string coff, bool full, bool deploy, bool replace);
+    LibRename(std::string pe, bool full, bool deploy, bool replace);
     bool ExecuteRename();
     bool ExecuteLibRename();
     bool ExecutePERename();
-    int ComputeDefFile();
+    bool ComputeDefFile();
     std::string ComputeRenameLink();
     std::string ComputeDefLine();
 
@@ -233,14 +246,14 @@ private:
     ExecuteCommand def_executor;
     ExecuteCommand lib_executor;
     std::string pe;
-    std::string name;
+    std::string coff;
     std::string new_lib;
     std::string def_file;
+    std::string tmp_def_file;
     bool full;
     bool deploy;
     bool replace;
     bool is_exe;
-    bool report;
 };
 
 
diff --git a/test/calc.h b/test/calc.h
deleted file mode 100644
index d7d3975..0000000
--- a/test/calc.h
+++ /dev/null
@@ -1,14 +0,0 @@
-/**
- * Copyright Spack Project Developers. See COPYRIGHT file for details.
- *
- * SPDX-License-Identifier: (Apache-2.0 OR MIT)
- */
-#pragma once
-
-#ifdef CALC_EXPORTS
-#define CALC_API __declspec(dllexport)
-#else
-#define CALC_API __declspec(dllimport)
-#endif
-
-extern "C" CALC_API int add(int a, int b);
