diff --git a/src/ld.cxx b/src/ld.cxx
index ff9bdde..890f364 100644
--- a/src/ld.cxx
+++ b/src/ld.cxx
@@ -6,6 +6,7 @@
 #include "ld.h"
 #include <minwindef.h>
 #include <cstdio>
+#include <iostream>
 #include "coff_parser.h"
 #include "coff_reader_writer.h"
 #include "linker_invocation.h"
@@ -34,7 +35,14 @@ DWORD LdInvocation::InvokeToolchain() {
     // We're creating a dll, we need to create an appropriate import lib
     if (!link_run.IsExeLink()) {
         std::string const imp_lib_name = link_run.get_implib_name();
-        std::string dll_name = link_run.get_mangled_out();
+        std::string dll_name;
+        try {
+            dll_name = link_run.get_mangled_out();
+        } catch (const SpackCompilerWrapperError& e) {
+            std::cerr << "Unable to mangle PE " << link_run.get_out()
+                      << " name is too long\n";
+            return ExitConditions::NORMALIZE_NAME_FAILURE;
+        }
         std::string const abs_out_imp_lib_name = imp_lib_name + ".dll-abs.lib";
         std::string def = "-def ";
         std::string piped_args = link_run.get_lib_link_args();
diff --git a/src/main.cxx b/src/main.cxx
index 9458d17..26d940d 100644
--- a/src/main.cxx
+++ b/src/main.cxx
@@ -89,12 +89,19 @@ int main(int argc, const char* argv[]) {
         }
         DEBUG = debug;
         std::unique_ptr<LibRename> rpath_lib;
-        if (has_coff) {
-            rpath_lib = std::make_unique<LibRename>(
-                patch_args.at("pe"), patch_args.at("coff"), full, deploy, true);
-        } else {
-            rpath_lib = std::make_unique<LibRename>(patch_args.at("pe"), full,
-                                                    deploy, true);
+        try {
+            if (has_coff) {
+                rpath_lib = std::make_unique<LibRename>(patch_args.at("pe"),
+                                                        patch_args.at("coff"),
+                                                        full, deploy, true);
+            } else {
+                rpath_lib = std::make_unique<LibRename>(patch_args.at("pe"),
+                                                        full, deploy, true);
+            }
+        } catch (const SpackCompilerWrapperError& e) {
+            std::cerr << "Cannot Rename PE file " << patch_args.at("pe")
+                      << " it contains references that are too long.\n";
+            return ExitConditions::RENAME_FAILURE;
         }
         if (!rpath_lib->ExecuteRename()) {
             std::cerr << "Library rename failed\n";
@@ -110,9 +117,17 @@ int main(int argc, const char* argv[]) {
             return ExitConditions::CLI_FAILURE;
         }
         if (report_args.find("pe") != report_args.end()) {
-            LibRename portable_executable(report_args.at("pe"), std::string(),
-                                          false, false, true);
-            portable_executable.ExecuteRename();
+            try {
+                LibRename portable_executable(
+                    report_args.at("pe"), std::string(), false, false, true);
+                portable_executable.ExecuteRename();
+            } catch (const SpackCompilerWrapperError& e) {
+                std::cerr
+                    << "Unable to parse command line for reporting\n"
+                    << "run command with --help flag for accepted command "
+                       "line arguments\n";
+                return ExitConditions::CLI_FAILURE;
+            }
         } else {
             CoffReaderWriter coff_reader(report_args.at("coff"));
             CoffParser coff(&coff_reader);
diff --git a/src/utils.cxx b/src/utils.cxx
index a22dcd5..5258caa 100644
--- a/src/utils.cxx
+++ b/src/utils.cxx
@@ -11,6 +11,7 @@
 #include <minwindef.h>
 #include <processenv.h>
 #include <stringapiset.h>
+#include <winbase.h>
 #include <winerror.h>
 #include <winnls.h>
 #include <winnt.h>
@@ -20,6 +21,7 @@
 #include <cstdlib>
 #include <cstring>
 #include <cwchar>
+#include <exception>
 #include <iostream>
 #include <limits>
 #include <map>
@@ -484,6 +486,12 @@ void replace_path_characters(char* path, size_t len) {
  * \param bsize the lengh of the padding to add
  */
 char* pad_path(const char* pth, DWORD str_size, DWORD bsize) {
+    // If str_size > bsize we get inappropriate conversion
+    // from signed to unsigned
+    if (str_size > bsize) {
+        debug("Padding string is greater than max string size allowed");
+        return nullptr;
+    }
     size_t const extended_buf = bsize - str_size + 2;
     char* padded_path = new char[bsize + 1];
     for (DWORD i = 0, j = 0; i < bsize && j < str_size; ++i) {
@@ -515,17 +523,16 @@ int get_padding_length(const std::string& name) {
     return count;
 }
 
-std::string strip_padding(const std::string &lib)
-{
+std::string strip_padding(const std::string& lib) {
     // One of the padding characters is a legitimate
     // path separator
-    int pad_len = get_padding_length(lib)-1;
+    int const pad_len = get_padding_length(lib) - 1;
     // Capture the drive and drive separator
-    std::string::const_iterator p = lib.cbegin();
-    std::string::const_iterator e = lib.cbegin()+2;
-    std::string stripped_drive(p, e);
+    std::string::const_iterator const p = lib.cbegin();
+    std::string::const_iterator e = lib.cbegin() + 2;
+    std::string const stripped_drive(p, e);
     e = e + pad_len;
-    std::string path_remainder(e, lib.end());
+    std::string const path_remainder(e, lib.end());
     return stripped_drive + path_remainder;
 }
 
@@ -545,6 +552,36 @@ std::string mangle_name(const std::string& name) {
         // relative paths, assume they're relative to the CWD of the linker (as they have to be)
         abs_out = join({GetCWD(), name}, "\\");
     }
+    // Now that we have the full path, check size
+    if (abs_out.length() > MAX_NAME_LEN) {
+        // Name is too long we need to attempt to shorten
+        // strip prefix
+        std::string pre = GetSpackEnv("SPACK_STAGE_DIR");
+        if (pre.empty()) {
+            pre = GetCWD();
+        }
+        std::string const rel = R"(\\?\)" + lstrip(abs_out, pre);
+        // Get SFN length so we can create buffer
+        DWORD const sfn_size = GetShortPathNameA(rel.c_str(), nullptr, 0);
+        char* sfn = new char[sfn_size + 1];
+        GetShortPathNameA(rel.c_str(), sfn, rel.length());
+        // sfn is null terminated per win32 api
+        std::string const rel_sfn = std::string(sfn);
+        std::string const new_abs_out = join({pre, rel_sfn}, "\\");
+        // If new, shortened path is too long, bail
+        if (new_abs_out.length() > MAX_NAME_LEN) {
+            std::cerr << "DLL path " << abs_out << " too long to relocate.\n";
+            std::cerr << "Shortened DLL path " << new_abs_out
+                      << " also too long to relocate.\n";
+            std::cerr << "Please move prefix " << pre
+                      << " to a shorter directory.\n";
+            delete[] sfn;
+            throw SpackCompilerWrapperError(
+                "DLL Path too long, cannot be relocated.");
+        }
+        delete[] sfn;
+        abs_out = new_abs_out;
+    }
     char* chr_abs_out = new char[abs_out.length() + 1];
     strcpy(chr_abs_out, abs_out.c_str());
     replace_path_characters(chr_abs_out, abs_out.length());
@@ -579,7 +616,7 @@ bool SpackInstalledLib(const std::string& lib) {
             "unset");
         return false;
     }
-    std::string stripped_lib = strip_padding(lib);
+    std::string const stripped_lib = strip_padding(lib);
     startswith(stripped_lib, prefix);
 }
 
@@ -732,3 +769,10 @@ char* findstr(char* search_str, const char* substr, size_t size) {
     }
     return nullptr;
 }
+
+SpackCompilerWrapperError::SpackCompilerWrapperError(char const* const message)
+    : std::runtime_error(message) {}
+
+char const* SpackCompilerWrapperError::what() const {
+    return exception::what();
+}
\ No newline at end of file
diff --git a/src/utils.h b/src/utils.h
index 83e0358..b692e38 100644
--- a/src/utils.h
+++ b/src/utils.h
@@ -172,7 +172,7 @@ void replace_path_characters(char* path, size_t len);
 
 void replace_special_characters(char* mangled, size_t len);
 
-bool SpackInstalledLib(const std::string &lib);
+bool SpackInstalledLib(const std::string& lib);
 
 // File and File handle helpers //
 
@@ -237,4 +237,10 @@ const std::map<char, char> path_to_special_characters{{'\\', '|'},
                                                       {'/', '|'},
                                                       {':', ';'}};
 
+class SpackCompilerWrapperError : public std::runtime_error {
+   public:
+    SpackCompilerWrapperError(char const* const message);
+    virtual char const* what() const;
+};
+
 static bool DEBUG = false;
diff --git a/src/winrpath.cxx b/src/winrpath.cxx
index 387b245..5671515 100644
--- a/src/winrpath.cxx
+++ b/src/winrpath.cxx
@@ -4,7 +4,7 @@
  * SPDX-License-Identifier: (Apache-2.0 OR MIT)
  */
 #include <cstdio>
-#include <stdio.h>
+#include <cstdio>
 #include <windows.h>  // NOLINT
 #include "winrpath.h"
 #include <fileapi.h>
@@ -102,7 +102,9 @@ bool LibRename::RenameDll(char* name_loc, const std::string& dll_path) const {
         char* new_lib_pth =
             pad_path(new_library_loc.c_str(),
                      static_cast<DWORD>(new_library_loc.size()));
-
+        if (!new_lib_pth) {
+            return false;
+        }
         replace_special_characters(new_lib_pth, MAX_NAME_LEN);
 
         // c_str returns a proper (i.e. null terminated) value, so we dont need to worry about
@@ -404,11 +406,18 @@ bool LibRename::ExecuteLibRename() {
         std::cerr << err;
         return false;
     }
-    std::string mangled_name = mangle_name(this->pe);
-    if (!coff_parser.NormalizeName(mangled_name)) {
-        std::cerr << "Unable to normalize name: " << mangled_name << "\n";
+    try {
+        std::string mangled_name = mangle_name(this->pe);
+        if (!coff_parser.NormalizeName(mangled_name)) {
+            std::cerr << "Unable to normalize name: " << mangled_name << "\n";
+            return false;
+        }
+    } catch (const SpackCompilerWrapperError& e) {
+        std::cerr << "Unable to mangle name, DLL name " << this->pe
+                  << " too long\n";
         return false;
     }
+
     return true;
 }
 
