diff --git a/.github/workflows/ci.yaml b/.github/workflows/ci.yaml
index cbc9f76..9979ea1 100644
--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -26,6 +26,9 @@ jobs:
       - name: Remove Git Bash tools
         run: |
           rmdir /s /q "C:\Program Files\Git\usr"
+      - name: enable 8.3 names
+        run: |
+          fsutil 8dot3name set 0
       - name: "Test RPath"
         run: |
           test\setup_and_drive_test.bat
diff --git a/Makefile b/Makefile
index 6a0f625..a86ace7 100644
--- a/Makefile
+++ b/Makefile
@@ -141,10 +141,41 @@ test_zerowrite: build_zerowrite_test
 	cl /c EHsc "test\src file\calc.cxx"
 	set SPACK_CC=%SPACK_CC_TMP%
 
+test_long_paths: build_and_check_test_sample
+	mkdir tmp\tmp\verylongdirectoryname\evenlongersubdirectoryname
+	xcopy /E test\include tmp\tmp\verylongdirectoryname\evenlongersubdirectoryname
+	xcopy /E "test\src file" tmp\tmp\verylongdirectoryname\evenlongersubdirectoryname
+	xcopy test\main.cxx tmp\tmp\verylongdirectoryname\evenlongersubdirectoryname
+	cd tmp\tmp\verylongdirectoryname\evenlongersubdirectoryname
+	rename calc.cxx verylongfilepathnamethatwillmostlikelybegreaterthanonehundredandfourtyfourcharacters.cxx
+	copy ..\..\..\..\cl.exe cl.exe
+	-@ if NOT EXIST "link.exe" mklink link.exe cl.exe
+	cl /c /EHsc "verylongfilepathnamethatwillmostlikelybegreaterthanonehundredandfourtyfourcharacters.cxx" /DCALC_EXPORTS /DCALC_HEADER="\"calc header/calc.h\"" /I include
+	cl /c /EHsc main.cxx /I include
+	link $(LFLAGS) verylongfilepathnamethatwillmostlikelybegreaterthanonehundredandfourtyfourcharacters.obj /DLL
+	link $(LFLAGS) main.obj verylongfilepathnamethatwillmostlikelybegreaterthanonehundredandfourtyfourcharacters.lib /out:tester.exe
+	tester.exe
+	cd ../../../..
+
+test_relocate_long_paths: test_long_paths
+	cd tmp\tmp\verylongdirectoryname\evenlongersubdirectoryname
+	-@ if NOT EXIST "relocate.exe" mklink relocate.exe cl.exe
+	cd ..
+	mkdir tmp_bin
+	mkdir tmp_lib
+	move evenlongersubdirectoryname\verylongfilepathnamethatwillmostlikelybegreaterthanonehundredandfourtyfourcharacters.dll tmp_bin\verylongfilepathnamethatwillmostlikelybegreaterthanonehundredandfourtyfourcharacters.dll
+	move evenlongersubdirectoryname\verylongfilepathnamethatwillmostlikelybegreaterthanonehundredandfourtyfourcharacters.lib tmp_lib\verylongfilepathnamethatwillmostlikelybegreaterthanonehundredandfourtyfourcharacters.lib
+	evenlongersubdirectoryname\relocate.exe --pe tmp_bin\verylongfilepathnamethatwillmostlikelybegreaterthanonehundredandfourtyfourcharacters.dll --coff tmp_lib\verylongfilepathnamethatwillmostlikelybegreaterthanonehundredandfourtyfourcharacters.lib --export
+	cd evenlongersubdirectoryname
+	del tester.exe
+	link main.obj ..\tmp_lib\verylongfilepathnamethatwillmostlikelybegreaterthanonehundredandfourtyfourcharacters.lib /out:tester.exe
+	.\tester.exe
+	cd ../../../..
+
 test_and_cleanup: test clean-test
 
 
-test: test_wrapper test_relocate_exe test_relocate_dll test_pipe_overflow
+test: test_wrapper test_relocate_exe test_relocate_dll test_long_paths test_pipe_overflow
 
 
 clean : clean-test clean-cl
diff --git a/src/coff_parser.cxx b/src/coff_parser.cxx
index f6b273c..3cda566 100644
--- a/src/coff_parser.cxx
+++ b/src/coff_parser.cxx
@@ -378,7 +378,8 @@ bool CoffParser::NormalizeName(std::string& name) {
     // The dll is found with and without an extenion, depending on the context of the location
     // i.e. in the section data, it can be found with both an extension and extensionless
     //  whereas in the symbol table or linker member strings, it's always found without an extension
-    std::string const name_no_ext = strip(name, ".dll");
+    std::string const name_no_dll = strip(name, ".dll");
+    std::string const name_no_ext = strip(name_no_dll, ".DLL");
     // Flag allowing us to skip multiple attempts
     // to rename the long names member this name
     bool long_name_renamed = false;
diff --git a/src/ld.cxx b/src/ld.cxx
index ff9bdde..890f364 100644
--- a/src/ld.cxx
+++ b/src/ld.cxx
@@ -6,6 +6,7 @@
 #include "ld.h"
 #include <minwindef.h>
 #include <cstdio>
+#include <iostream>
 #include "coff_parser.h"
 #include "coff_reader_writer.h"
 #include "linker_invocation.h"
@@ -34,7 +35,14 @@ DWORD LdInvocation::InvokeToolchain() {
     // We're creating a dll, we need to create an appropriate import lib
     if (!link_run.IsExeLink()) {
         std::string const imp_lib_name = link_run.get_implib_name();
-        std::string dll_name = link_run.get_mangled_out();
+        std::string dll_name;
+        try {
+            dll_name = link_run.get_mangled_out();
+        } catch (const SpackCompilerWrapperError& e) {
+            std::cerr << "Unable to mangle PE " << link_run.get_out()
+                      << " name is too long\n";
+            return ExitConditions::NORMALIZE_NAME_FAILURE;
+        }
         std::string const abs_out_imp_lib_name = imp_lib_name + ".dll-abs.lib";
         std::string def = "-def ";
         std::string piped_args = link_run.get_lib_link_args();
diff --git a/src/linker_invocation.cxx b/src/linker_invocation.cxx
index 5e7b60b..fc6316f 100644
--- a/src/linker_invocation.cxx
+++ b/src/linker_invocation.cxx
@@ -50,7 +50,10 @@ void LinkerInvocation::Parse() {
             this->is_exe_ = false;
         } else if (startswith(normal_token, "-out") ||
                    startswith(normal_token, "/out")) {
-            this->output_ = split(*token, ":")[1];
+            StrList split_out = split(*token, ":", 1);
+            StrList const split_name =
+                StrList(split_out.begin() + 1, split_out.end());
+            this->output_ = join(split_name, "\\");
         } else if (endswith(normal_token, ".obj")) {
             this->objs_.push_back(*token);
         } else if (startswith(normal_token, "@") &&
@@ -67,7 +70,11 @@ void LinkerInvocation::Parse() {
     }
     std::string const ext = this->is_exe_ ? ".exe" : ".dll";
     if (this->output_.empty()) {
-        this->output_ = strip(this->objs_.front(), ".obj") + ext;
+        // with no "out" argument, the linker
+        // will place the file in the CWD
+        std::string const name_obj = this->objs_.front();
+        std::string const filename = split(name_obj, "\\").back();
+        this->output_ = join({GetCWD(), strip(filename, ".obj")}, "\\") + ext;
     }
     this->name_ = strip(this->output_, ext);
     if (this->implibname_.empty()) {
diff --git a/src/main.cxx b/src/main.cxx
index 9458d17..26d940d 100644
--- a/src/main.cxx
+++ b/src/main.cxx
@@ -89,12 +89,19 @@ int main(int argc, const char* argv[]) {
         }
         DEBUG = debug;
         std::unique_ptr<LibRename> rpath_lib;
-        if (has_coff) {
-            rpath_lib = std::make_unique<LibRename>(
-                patch_args.at("pe"), patch_args.at("coff"), full, deploy, true);
-        } else {
-            rpath_lib = std::make_unique<LibRename>(patch_args.at("pe"), full,
-                                                    deploy, true);
+        try {
+            if (has_coff) {
+                rpath_lib = std::make_unique<LibRename>(patch_args.at("pe"),
+                                                        patch_args.at("coff"),
+                                                        full, deploy, true);
+            } else {
+                rpath_lib = std::make_unique<LibRename>(patch_args.at("pe"),
+                                                        full, deploy, true);
+            }
+        } catch (const SpackCompilerWrapperError& e) {
+            std::cerr << "Cannot Rename PE file " << patch_args.at("pe")
+                      << " it contains references that are too long.\n";
+            return ExitConditions::RENAME_FAILURE;
         }
         if (!rpath_lib->ExecuteRename()) {
             std::cerr << "Library rename failed\n";
@@ -110,9 +117,17 @@ int main(int argc, const char* argv[]) {
             return ExitConditions::CLI_FAILURE;
         }
         if (report_args.find("pe") != report_args.end()) {
-            LibRename portable_executable(report_args.at("pe"), std::string(),
-                                          false, false, true);
-            portable_executable.ExecuteRename();
+            try {
+                LibRename portable_executable(
+                    report_args.at("pe"), std::string(), false, false, true);
+                portable_executable.ExecuteRename();
+            } catch (const SpackCompilerWrapperError& e) {
+                std::cerr
+                    << "Unable to parse command line for reporting\n"
+                    << "run command with --help flag for accepted command "
+                       "line arguments\n";
+                return ExitConditions::CLI_FAILURE;
+            }
         } else {
             CoffReaderWriter coff_reader(report_args.at("coff"));
             CoffParser coff(&coff_reader);
diff --git a/src/utils.cxx b/src/utils.cxx
index a22dcd5..eb1ce9c 100644
--- a/src/utils.cxx
+++ b/src/utils.cxx
@@ -11,15 +11,18 @@
 #include <minwindef.h>
 #include <processenv.h>
 #include <stringapiset.h>
+#include <winbase.h>
 #include <winerror.h>
 #include <winnls.h>
 #include <winnt.h>
+#include <winsock.h>
 
 #include <algorithm>
 #include <cctype>
 #include <cstdlib>
 #include <cstring>
 #include <cwchar>
+#include <exception>
 #include <iostream>
 #include <limits>
 #include <map>
@@ -122,16 +125,25 @@ std::wstring ConvertASCIIToWide(const std::string& str) {
  * Decomposes the input string into a list separated by
  * delim
  * 
+ * Count determines how many delims will be processed
+ * if count > 0
+ * 
  * Returns the list produced by breaking up input string s on delim
  */
-StrList split(const std::string& str, const std::string& delim) {
+StrList split(const std::string& str, const std::string& delim,
+              const u_int count) {
     size_t pos_start = 0;
     size_t pos_end;
     size_t const delim_len = delim.length();
     std::string token;
     StrList res = StrList();
-
-    while ((pos_end = str.find(delim, pos_start)) != std::string::npos) {
+    u_int delim_count = count;
+    if (!count) {
+        delim_count += 1;
+    }
+    u_int delim_found = 0;
+    while (((pos_end = str.find(delim, pos_start)) != std::string::npos) &&
+           delim_found < delim_count) {
         size_t const token_len = pos_end - pos_start;
         token = str.substr(pos_start, token_len);
         pos_start = pos_end + delim_len;
@@ -139,6 +151,9 @@ StrList split(const std::string& str, const std::string& delim) {
             continue;
         }
         res.push_back(token);
+        if (count) {
+            delim_found += 1;
+        }
     }
     res.push_back(str.substr(pos_start));
     return res;
@@ -163,7 +178,7 @@ std::string strip(const std::string& str, const std::string& substr) {
 std::string lstrip(const std::string& str, const std::string& substr) {
     if (!startswith(str, substr))
         return str;
-    return str.substr(substr.size() - 1, str.size());
+    return str.substr(substr.size(), str.size());
 }
 
 /**
@@ -484,6 +499,12 @@ void replace_path_characters(char* path, size_t len) {
  * \param bsize the lengh of the padding to add
  */
 char* pad_path(const char* pth, DWORD str_size, DWORD bsize) {
+    // If str_size > bsize we get inappropriate conversion
+    // from signed to unsigned
+    if (str_size > bsize) {
+        debug("Padding string is greater than max string size allowed");
+        return nullptr;
+    }
     size_t const extended_buf = bsize - str_size + 2;
     char* padded_path = new char[bsize + 1];
     for (DWORD i = 0, j = 0; i < bsize && j < str_size; ++i) {
@@ -515,20 +536,65 @@ int get_padding_length(const std::string& name) {
     return count;
 }
 
-std::string strip_padding(const std::string &lib)
-{
+std::string strip_padding(const std::string& lib) {
     // One of the padding characters is a legitimate
     // path separator
-    int pad_len = get_padding_length(lib)-1;
+    int const pad_len = get_padding_length(lib) - 1;
     // Capture the drive and drive separator
-    std::string::const_iterator p = lib.cbegin();
-    std::string::const_iterator e = lib.cbegin()+2;
-    std::string stripped_drive(p, e);
+    std::string::const_iterator const p = lib.cbegin();
+    std::string::const_iterator e = lib.cbegin() + 2;
+    std::string const stripped_drive(p, e);
     e = e + pad_len;
-    std::string path_remainder(e, lib.end());
+    std::string const path_remainder(e, lib.end());
     return stripped_drive + path_remainder;
 }
 
+std::string getSFN(const std::string& path) {
+    std::string const escaped = R"(\\?\)" + path;
+    // Get SFN length so we can create buffer
+    DWORD const sfn_size =
+        GetShortPathNameA(escaped.c_str(), NULL, 0);  //NOLINT
+    char* sfn = new char[sfn_size + 1];
+    GetShortPathNameA(escaped.c_str(), sfn, escaped.length());
+    // sfn is null terminated per win32 api
+    std::string s_sfn = std::string(sfn);
+    delete[] sfn;
+    return s_sfn;
+}
+
+/**
+ * Replace path after a given prefix with the SFN representation
+ */
+std::string short_name_post_prefix(const std::string& path) {
+    // strip prefix
+    std::string pre = GetSpackEnv("SPACK_CONTEXT_ROOT");
+    if (pre.empty()) {
+        pre = GetCWD();
+    }
+    // Get SFN for path to name
+    // Use "disable string parsing" prefix in case
+    // the path is too long
+    std::string const abs_sfn = getSFN(path);
+    // Get SFN for path to name prefix
+    std::string const pre_sfn = getSFN(pre);
+    // Strip prefix SFN so we can prepend the real prefix to it
+    // Prefix should be spack stage root, which allows us the maximal
+    // possible path shortening without effecting potential naming collisions
+    // and still allowing us to determine we're in a spack context
+    std::string const rel_sfn = lstrip(abs_sfn, pre_sfn);
+    std::string new_abs_out = pre + rel_sfn;
+    if (new_abs_out.length() > MAX_NAME_LEN) {
+        std::cerr << "DLL path " << path << " too long to relocate.\n";
+        std::cerr << "Shortened DLL path " << new_abs_out
+                  << " also too long to relocate.\n";
+        std::cerr << "Please move Spack prefix "
+                  << " to a shorter directory.\n";
+        throw SpackCompilerWrapperError(
+            "DLL Path too long, cannot be relocated.");
+    }
+    return new_abs_out;
+}
+
 /**
  * Mangles a string representing a path to have no path characters
  *  instead path characters (i.e. \\, :, etc) are replaced with
@@ -545,6 +611,13 @@ std::string mangle_name(const std::string& name) {
         // relative paths, assume they're relative to the CWD of the linker (as they have to be)
         abs_out = join({GetCWD(), name}, "\\");
     }
+    // Now that we have the full path, check size
+    if (abs_out.length() > MAX_NAME_LEN) {
+        // Name is too long we need to attempt to shorten
+        std::string const new_abs_out = short_name_post_prefix(abs_out);
+        // If new, shortened path is too long, bail
+        abs_out = new_abs_out;
+    }
     char* chr_abs_out = new char[abs_out.length() + 1];
     strcpy(chr_abs_out, abs_out.c_str());
     replace_path_characters(chr_abs_out, abs_out.length());
@@ -579,7 +652,7 @@ bool SpackInstalledLib(const std::string& lib) {
             "unset");
         return false;
     }
-    std::string stripped_lib = strip_padding(lib);
+    std::string const stripped_lib = strip_padding(lib);
     startswith(stripped_lib, prefix);
 }
 
@@ -732,3 +805,10 @@ char* findstr(char* search_str, const char* substr, size_t size) {
     }
     return nullptr;
 }
+
+SpackCompilerWrapperError::SpackCompilerWrapperError(char const* const message)
+    : std::runtime_error(message) {}
+
+char const* SpackCompilerWrapperError::what() const {
+    return exception::what();
+}
\ No newline at end of file
diff --git a/src/utils.h b/src/utils.h
index 83e0358..eaa691b 100644
--- a/src/utils.h
+++ b/src/utils.h
@@ -71,7 +71,8 @@ std::wstring ConvertASCIIToWide(const std::string& str);
 // Splits argument "s" by delineator delim
 // Returns vector of strings, if delim is present
 // Returns a single item list
-StrList split(const std::string& s, const std::string& delim);
+StrList split(const std::string& s, const std::string& delim,
+              const u_int count = 0);
 
 //Strips substr off the RHS of the larger string
 std::string strip(const std::string& s, const std::string& substr);
@@ -162,6 +163,8 @@ bool IsPathAbsolute(const std::string& pth);
 
 bool hasPathCharacters(const std::string& name);
 
+std::string short_name_post_prefix(const std::string& path);
+
 std::string mangle_name(const std::string& name);
 
 int get_padding_length(const std::string& name);
@@ -172,7 +175,7 @@ void replace_path_characters(char* path, size_t len);
 
 void replace_special_characters(char* mangled, size_t len);
 
-bool SpackInstalledLib(const std::string &lib);
+bool SpackInstalledLib(const std::string& lib);
 
 // File and File handle helpers //
 
@@ -237,4 +240,10 @@ const std::map<char, char> path_to_special_characters{{'\\', '|'},
                                                       {'/', '|'},
                                                       {':', ';'}};
 
+class SpackCompilerWrapperError : public std::runtime_error {
+   public:
+    SpackCompilerWrapperError(char const* const message);
+    virtual char const* what() const;
+};
+
 static bool DEBUG = false;
diff --git a/src/winrpath.cxx b/src/winrpath.cxx
index 387b245..fd3c163 100644
--- a/src/winrpath.cxx
+++ b/src/winrpath.cxx
@@ -92,17 +92,28 @@ bool LibRename::RenameDll(char* name_loc, const std::string& dll_path) const {
             return false;
         }
         LibraryFinder lib_finder;
-        std::string const new_library_loc =
+        std::string new_library_loc =
             lib_finder.FindLibrary(file_name, dll_path);
         if (new_library_loc.empty()) {
             std::cerr << "Unable to find library " << file_name << " from "
                       << dll_path << " for relocation" << "\n";
             return false;
         }
+        if (new_library_loc.length() > MAX_NAME_LEN) {
+            try {
+                std::string const short_lib_loc =
+                    short_name_post_prefix(new_library_loc);
+                new_library_loc = short_lib_loc;
+            } catch (SpackCompilerWrapperError& e) {
+                return false;
+            }
+        }
         char* new_lib_pth =
             pad_path(new_library_loc.c_str(),
                      static_cast<DWORD>(new_library_loc.size()));
-
+        if (!new_lib_pth) {
+            return false;
+        }
         replace_special_characters(new_lib_pth, MAX_NAME_LEN);
 
         // c_str returns a proper (i.e. null terminated) value, so we dont need to worry about
@@ -404,11 +415,18 @@ bool LibRename::ExecuteLibRename() {
         std::cerr << err;
         return false;
     }
-    std::string mangled_name = mangle_name(this->pe);
-    if (!coff_parser.NormalizeName(mangled_name)) {
-        std::cerr << "Unable to normalize name: " << mangled_name << "\n";
+    try {
+        std::string mangled_name = mangle_name(this->pe);
+        if (!coff_parser.NormalizeName(mangled_name)) {
+            std::cerr << "Unable to normalize name: " << mangled_name << "\n";
+            return false;
+        }
+    } catch (const SpackCompilerWrapperError& e) {
+        std::cerr << "Unable to mangle name, DLL name " << this->pe
+                  << " too long\n";
         return false;
     }
+
     return true;
 }
 
