diff --git a/Makefile b/Makefile
index c90ccce..31fcd5f 100644
--- a/Makefile
+++ b/Makefile
@@ -22,7 +22,7 @@ PREFIX="$(MAKEDIR)\install\"
 !ENDIF
 
 !IF "$(BUILD_TYPE)" == "DEBUG"
-BUILD_CFLAGS = /Zi /fsanitize=address
+BUILD_CFLAGS = /Zi
 BUILD_LINK = /DEBUG
 !ENDIF
 
@@ -59,7 +59,7 @@ setup_test: cl.exe
 # smoke test - can the wrapper compile anything
 build_and_check_test_sample : setup_test
 	cd tmp\test
-	cl /c /EHsc ..\..\test\calc.cxx /DCALC_EXPORTS /I ..\..\test\include
+	cl /c /EHsc "..\..\test\src file\calc.cxx" /DCALC_EXPORTS /DCALC_HEADER=\"calc.h\" /I ..\..\test\include
 	cl /c /EHsc ..\..\test\main.cxx /I ..\..\test\include
 	link $(LFLAGS) calc.obj /out:calc.dll /DLL
 	link $(LFLAGS) main.obj calc.lib /out:tester.exe
diff --git a/src/utils.cxx b/src/utils.cxx
index 8e153b0..220be6f 100644
--- a/src/utils.cxx
+++ b/src/utils.cxx
@@ -188,6 +188,13 @@ std::string quoteAsNeeded(std::string &str) {
         // There are spaces or special characters in string, quote it
         return "\"" + str + "\"";
     }
+    if (str.find_first_of("\"") != std::string::npos) {
+        // If there are escaped quotes in input
+        // We need to escape them as well as we're adding another
+        // layer of indirection between builder and compiler
+        std::regex pattern("\"");
+        return std::regex_replace(str, pattern, "\\\"");
+    }
     return str;
 }
 
diff --git a/test/calc.cxx b/test/src file/calc.cxx
similarity index 73%
rename from test/calc.cxx
rename to test/src file/calc.cxx
index cd7a136..b07cca2 100644
--- a/test/calc.cxx
+++ b/test/src file/calc.cxx	
@@ -3,7 +3,12 @@
  *
  * SPDX-License-Identifier: (Apache-2.0 OR MIT)
  */
+
+#if defined(CALC_HEADER)
+#include CALC_HEADER /* "calc.h" */
+#else
 #include "calc.h"
+#endif
 
 extern "C" int add(int a, int b)
 {
