diff --git a/src/utils.cxx b/src/utils.cxx
index 764dda6..9f52350 100644
--- a/src/utils.cxx
+++ b/src/utils.cxx
@@ -490,6 +490,20 @@ int get_padding_length(const std::string &name)
     return c;
 }
 
+std::string strip_padding(const std::string &lib)
+{
+    // One of the padding characters is a legitimate
+    // path separator
+    int pad_len = get_padding_length(lib)-1;
+    // Capture the drive and drive separator
+    std::string::const_iterator p = lib.cbegin();
+    std::string::const_iterator e = lib.cbegin()+2;
+    std::string stripped_drive(p, e);
+    e = e + pad_len;
+    std::string path_remainder(e, lib.end());
+    return stripped_drive + path_remainder;
+}
+
 /**
  * Mangles a string representing a path to have no path characters
  *  instead path characters (i.e. \\, :, etc) are replaced with
@@ -540,7 +554,8 @@ bool SpackInstalledLib(const std::string &lib) {
         debug("Unable to determine Spack install prefix, SPACK_INSTALL_PREFIX unset");
         return false;
     }
-    return startswith(lib, prefix);
+    std::string stripped_lib = strip_padding(lib);
+    startswith(stripped_lib, prefix);
 }
 
 LibraryFinder::LibraryFinder() : search_vars{"SPACK_RELOCATE_PATH"} {}
