diff --git a/third_party/systemlibs/protobuf.BUILD b/third_party/systemlibs/protobuf.BUILD
--- a/third_party/systemlibs/protobuf.BUILD
+++ b/third_party/systemlibs/protobuf.BUILD
@@ -108,6 +108,30 @@
 [proto_library(
     name = proto[0] + "_proto",
     srcs = [proto[1][0]],
     visibility = ["//visibility:public"],
     deps = [dep + "_proto" for dep in proto[1][1]],
 ) for proto in WELL_KNOWN_PROTO_MAP.items()]
+
+cc_library(
+    name = "protobuf_lite",
+    linkopts = ["-lprotobuf-lite"],
+    visibility = ["//visibility:public"],
+)
+
+filegroup(
+    name = "well_known_protos",
+    srcs = [":link_proto_files"],
+    visibility = ["//visibility:public"],
+)
+
+cc_library(
+    name = "proto_api",
+    linkopts = ["-lprotobuf"],
+    visibility = ["//visibility:public"],
+)
+
+py_proto_library(
+    name = "well_known_types_py_pb2",
+    srcs = RELATIVE_WELL_KNOWN_PROTOS,
+    include = ".",
+    default_runtime = ":protobuf_python",
+    protoc = ":protoc",
+    visibility = ["//visibility:public"],
+)
