diff --git a/cuda/backprop/Makefile b/cuda/backprop/Makefile
index a7ed41d..a6812bd 100755
--- a/cuda/backprop/Makefile
+++ b/cuda/backprop/Makefile
@@ -22,7 +22,9 @@ endif
 
 
 backprop: backprop.o facetrain.o imagenet.o backprop_cuda.o 
-	$(CC) $(CC_FLAGS) backprop.o facetrain.o imagenet.o backprop_cuda.o -o backprop -L$(CUDA_LIB_DIR) -lcuda -lcudart -lm
+	$(NVCC) $(NVCC_FLAGS) \
+	   	backprop.o facetrain.o imagenet.o backprop_cuda.o -o backprop \
+		-lm -Xlinker -pie
 
 %.o: %.[ch]
 	$(CC) $(CC_FLAGS) $< -c
diff --git a/cuda/backprop/backprop.c b/cuda/backprop/backprop.c
index 0c6b95f..bb19c08 100755
--- a/cuda/backprop/backprop.c
+++ b/cuda/backprop/backprop.c
@@ -307,7 +307,7 @@ int nh, no;
 
 void bpnn_adjust_weights(delta, ndelta, ly, nly, w, oldw)
 float *delta, *ly, **w, **oldw;
-int ndelta;
+int ndelta, nly;
 {
   float new_dw;
   int k, j;
diff --git a/cuda/backprop/imagenet.c b/cuda/backprop/imagenet.c
index 2181fc9..8319bd3 100755
--- a/cuda/backprop/imagenet.c
+++ b/cuda/backprop/imagenet.c
@@ -3,7 +3,7 @@
 #include <stdlib.h>
 #include "backprop.h"
 
-extern layer_size;
+extern int layer_size;
 
 void load(net)
 BPNN *net;
diff --git a/cuda/lavaMD/main.c b/cuda/lavaMD/main.c
index b79bc58..6917f3a 100755
--- a/cuda/lavaMD/main.c
+++ b/cuda/lavaMD/main.c
@@ -20,6 +20,7 @@
 #include <stdio.h>					// (in path known to compiler)			needed by printf
 #include <stdlib.h>					// (in path known to compiler)			needed by malloc
 #include <stdbool.h>				// (in path known to compiler)			needed by true/false
+#include <string.h>
 
 //======================================================================================================================================================150
 //	UTILITIES
diff --git a/cuda/lavaMD/makefile b/cuda/lavaMD/makefile
index 3944253..3ef2813 100755
--- a/cuda/lavaMD/makefile
+++ b/cuda/lavaMD/makefile
@@ -26,14 +26,12 @@ a.out:		main.o \
 			./util/num/num.o \
 			./util/timer/timer.o \
 			./util/device/device.o
-	$(C_C)	$(KERNEL_DIM) main.o \
+	$(CUD_C)	$(KERNEL_DIM) main.o \
 			./kernel/kernel_gpu_cuda_wrapper.o \
 			./util/num/num.o \
 			./util/timer/timer.o \
 			./util/device/device.o \
-			-lm \
-			-L/usr/local/cuda/lib64 \
-			-lcuda -lcudart \
+			-lm -Xlinker -pie \
 			$(OMP_LIB) \
 			-o lavaMD
 
diff --git a/cuda/lavaMD/makefile.orig b/cuda/lavaMD/makefile.orig
new file mode 100755
index 0000000..3944253
--- /dev/null
+++ b/cuda/lavaMD/makefile.orig
@@ -0,0 +1,92 @@
+include ../../common/make.config
+
+# Example
+# target: dependencies
+	# command 1
+	# command 2
+          # .
+          # .
+          # .
+	# command n
+ifdef OUTPUT
+override OUTPUT = -DOUTPUT
+endif
+
+C_C = gcc
+OMP_LIB = -lgomp
+OMP_FLAG = -fopenmp
+
+CUD_C = $(CUDA_DIR)/bin/nvcc
+# OMP_FLAG = 	-Xcompiler paste_one_here
+CUDA_FLAG = -arch sm_13
+
+# link objects (binaries) together
+a.out:		main.o \
+			./kernel/kernel_gpu_cuda_wrapper.o \
+			./util/num/num.o \
+			./util/timer/timer.o \
+			./util/device/device.o
+	$(C_C)	$(KERNEL_DIM) main.o \
+			./kernel/kernel_gpu_cuda_wrapper.o \
+			./util/num/num.o \
+			./util/timer/timer.o \
+			./util/device/device.o \
+			-lm \
+			-L/usr/local/cuda/lib64 \
+			-lcuda -lcudart \
+			$(OMP_LIB) \
+			-o lavaMD
+
+# compile function files into objects (binaries)
+main.o:		main.h \
+			main.c \
+			./kernel/kernel_gpu_cuda_wrapper.h \
+			./kernel/kernel_gpu_cuda_wrapper.cu \
+			./util/num/num.h \
+			./util/num/num.c \
+			./util/timer/timer.h \
+			./util/timer/timer.c \
+			./util/device/device.h \
+			./util/device/device.cu
+	$(C_C)	$(KERNEL_DIM) $(OUTPUT) main.c \
+			-c \
+			-o main.o \
+			-O3
+
+./kernel/kernel_gpu_cuda_wrapper.o:	./kernel/kernel_gpu_cuda_wrapper.h \
+									./kernel/kernel_gpu_cuda_wrapper.cu
+	$(CUD_C) $(KERNEL_DIM)						./kernel/kernel_gpu_cuda_wrapper.cu \
+									-c \
+									-o ./kernel/kernel_gpu_cuda_wrapper.o \
+									-O3 \
+									$(CUDA_FLAG)
+
+./util/num/num.o:	./util/num/num.h \
+					./util/num/num.c
+	$(C_C)			./util/num/num.c \
+					-c \
+					-o ./util/num/num.o \
+					-O3
+
+./util/timer/timer.o:	./util/timer/timer.h \
+						./util/timer/timer.c
+	$(C_C)				./util/timer/timer.c \
+						-c \
+						-o ./util/timer/timer.o \
+						-O3
+
+./util/device/device.o:	./util/device/device.h \
+						./util/device/device.cu
+	$(CUD_C)			./util/device/device.cu \
+						-c \
+						-o ./util/device/device.o \
+						-O3
+
+# delete all object and executable files
+clean:
+	rm	*.o \
+		./kernel/*.o \
+		./util/num/*.o \
+		./util/timer/*.o \
+		./util/device/*.o \
+		lavaMD
diff --git a/cuda/lavaMD/makefile.rej b/cuda/lavaMD/makefile.rej
new file mode 100644
index 0000000..f770aa1
--- /dev/null
+++ b/cuda/lavaMD/makefile.rej
@@ -0,0 +1,19 @@
+--- cuda/lavaMD/makefile
++++ cuda/lavaMD/makefile
+@@ -26,14 +26,12 @@ a.out:		main.o \
+ 			./util/num/num.o \
+ 			./util/timer/timer.o \
+ 			./util/device/device.o
+-	$(C_C)	$(KERNEL_DIM) main.o \
++	$(CUD_C)	$(KERNEL_DIM) main.o \
+ 			./kernel/kernel_gpu_cuda_wrapper.o \
+ 			./util/num/num.o \
+ 			./util/timer/timer.o \
+ 			./util/device/device.o \
+- 			-lm \
+-			-L/usr/local/cuda/lib64 \
+-			-lcuda -lcudart \
++ 			-lm -Xlinker -pie \
+ 			$(OMP_LIB) \
+ 			-o lavaMD
+ 
diff --git a/cuda/lavaMD/util/timer/timer.c b/cuda/lavaMD/util/timer/timer.c
index c7cc252..28a13e1 100755
--- a/cuda/lavaMD/util/timer/timer.c
+++ b/cuda/lavaMD/util/timer/timer.c
@@ -1,3 +1,4 @@
+#include "./timer.h"
 #ifdef __cplusplus
 extern "C" {
 #endif
@@ -11,6 +12,7 @@ extern "C" {
 //======================================================================================================================================================150
 
 #include <stdlib.h>
+#include <sys/time.h>
 
 //======================================================================================================================================================150
 //	FUNCTIONS
diff --git a/cuda/lavaMD/util/timer/timer.h b/cuda/lavaMD/util/timer/timer.h
index 1744df4..c89ec8a 100755
--- a/cuda/lavaMD/util/timer/timer.h
+++ b/cuda/lavaMD/util/timer/timer.h
@@ -1,3 +1,5 @@
+#include <time.h>
+
 #ifdef __cplusplus
 extern "C" {
 #endif
