diff --git a/cuda/leukocyte/CUDA/Makefile b/cuda/leukocyte/CUDA/Makefile
index b4f6962..3fe998c 100755
--- a/cuda/leukocyte/CUDA/Makefile
+++ b/cuda/leukocyte/CUDA/Makefile
@@ -29,7 +29,9 @@ endif
 
 
 leukocyte: detect_main.o avilib.o find_ellipse.o find_ellipse_kernel.o track_ellipse.o track_ellipse_kernel.o misc_math.o $(MATRIX_DIR)/meschach.a
-	$(CC) $(CC_FLAGS) -lm avilib.o find_ellipse.o find_ellipse_kernel.o track_ellipse.o track_ellipse_kernel.o misc_math.o detect_main.o -o leukocyte $(MATRIX_DIR)/meschach.a -L$(CUDA_LIB_DIR) -lm -lcuda -lcudart
+	$(NVCC) $(NVCC_FLAGS) -lm \
+		avilib.o find_ellipse.o find_ellipse_kernel.o track_ellipse.o track_ellipse_kernel.o misc_math.o detect_main.o -o leukocyte \
+		$(MATRIX_DIR)/meschach.a -L$(CUDA_LIB_DIR) -lcuda
 
 %.o: %.[ch]
 	$(CC)  $(OUTPUT) $(CC_FLAGS) $< -c
diff --git a/cuda/leukocyte/CUDA/find_ellipse_kernel.cu b/cuda/leukocyte/CUDA/find_ellipse_kernel.cu
index 1a52152..ff4b1ee 100755
--- a/cuda/leukocyte/CUDA/find_ellipse_kernel.cu
+++ b/cuda/leukocyte/CUDA/find_ellipse_kernel.cu
@@ -25,8 +25,10 @@ __constant__ int c_tX[NCIRCLES * NPOINTS];
 __constant__ int c_tY[NCIRCLES * NPOINTS];
 
 // Texture references to the gradient matrices used by the GICOV kernel
-texture<float, 1, cudaReadModeElementType> t_grad_x;
-texture<float, 1, cudaReadModeElementType> t_grad_y;
+__device__ cudaTextureObject_t t_grad_x = 0;
+__device__ cudaTextureObject_t t_grad_y = 0;
+//texture<float, 1, cudaReadModeElementType> t_grad_x;
+//texture<float, 1, cudaReadModeElementType> t_grad_y;
 
 // Kernel to find the maximal GICOV value at each pixel of a
 //  video frame, based on the input x- and y-gradient matrices
@@ -54,8 +56,8 @@ __global__ void GICOV_kernel(int grad_m, float *gicov) {
 			
 			// Compute the combined gradient value at the current sample point
 			int addr = x * grad_m + y;
-			float p = tex1Dfetch(t_grad_x,addr) * c_cos_angle[n] + 
-					  tex1Dfetch(t_grad_y,addr) * c_sin_angle[n];
+			float p = tex1D<float>(t_grad_x,addr) * c_cos_angle[n] + 
+					  tex1D<float>(t_grad_y,addr) * c_sin_angle[n];
 			
 			// Update the running total
 			sum += p;
@@ -97,8 +99,34 @@ float *GICOV_CUDA(int grad_m, int grad_n, float *host_grad_x, float *host_grad_y
 	cudaMemcpy(device_grad_y, host_grad_y, grad_mem_size, cudaMemcpyHostToDevice);
     
 	// Bind the device arrays to texture references
-    cudaBindTexture(0, t_grad_x, device_grad_x, grad_mem_size);
-    cudaBindTexture(0, t_grad_y, device_grad_y, grad_mem_size);
+	cudaChannelFormatDesc chDescX = cudaCreateChannelDesc<float>();
+    cudaResourceDesc resDescX = {};
+    resDescX.resType = cudaResourceTypeLinear;
+	resDescX.res.linear.devPtr = device_grad_x;
+	resDescX.res.linear.desc = chDescX;
+	resDescX.res.linear.sizeInBytes = grad_mem_size;
+	cudaTextureDesc texDescX = {};
+    memset(&texDescX, 0, sizeof(texDescX));
+    texDescX.filterMode = cudaFilterModePoint;
+    texDescX.readMode = cudaReadModeElementType;
+    texDescX.normalizedCoords = 0;
+	cudaCreateTextureObject(&t_grad_x, &resDescX, &texDescX, NULL);
+
+	cudaChannelFormatDesc chDescY = cudaCreateChannelDesc<float>();
+    cudaResourceDesc resDescY = {};
+    resDescY.resType = cudaResourceTypeLinear;
+	resDescY.res.linear.devPtr = device_grad_y;
+	resDescY.res.linear.desc = chDescY;
+	resDescY.res.linear.sizeInBytes = grad_mem_size;
+	cudaTextureDesc texDescY = {};
+    memset(&texDescY, 0, sizeof(texDescY));
+    texDescY.filterMode = cudaFilterModePoint;
+    texDescY.readMode = cudaReadModeElementType;
+    texDescY.normalizedCoords = 0;
+	cudaCreateTextureObject(&t_grad_y, &resDescY, &texDescY, NULL);
+
+    //cudaBindTexture(0, t_grad_x, device_grad_x, grad_mem_size);
+    //cudaBindTexture(0, t_grad_y, device_grad_y, grad_mem_size);
 
 	// Allocate & initialize device memory for result
 	// (some elements are not assigned values in the kernel)
@@ -125,8 +153,10 @@ float *GICOV_CUDA(int grad_m, int grad_n, float *host_grad_x, float *host_grad_y
 	cudaMemcpy(host_gicov, device_gicov, grad_mem_size, cudaMemcpyDeviceToHost);
 
 	// Cleanup memory
-	cudaUnbindTexture(t_grad_x);
-	cudaUnbindTexture(t_grad_y);
+	//cudaUnbindTexture(t_grad_x);
+	//cudaUnbindTexture(t_grad_y);
+	cudaDestroyTextureObject(t_grad_x);
+	cudaDestroyTextureObject(t_grad_y);
 	cudaFree(device_grad_x);
 	cudaFree(device_grad_y);
 
@@ -138,7 +168,8 @@ float *GICOV_CUDA(int grad_m, int grad_n, float *host_grad_x, float *host_grad_y
 __constant__ float c_strel[STREL_SIZE * STREL_SIZE];
 
 // Texture reference to the GICOV matrix used by the dilation kernel
-texture<float, 1, cudaReadModeElementType> t_img;
+__device__ cudaTextureObject_t t_img = 0;
+//texture<float, 1, cudaReadModeElementType> t_img;
 
 // Kernel to compute the dilation of the GICOV matrix produced by the GICOV kernel
 // Each element (i, j) of the output matrix is set equal to the maximal value in
@@ -173,7 +204,7 @@ __global__ void dilate_kernel(int img_m, int img_n, int strel_m, int strel_n, fl
 					(c_strel[(el_i * strel_n) + el_j] != 0) ) {
 						// Determine if this is maximal value seen so far
 						int addr = (x * img_m) + y;
-						float temp = tex1Dfetch(t_img, addr);
+						float temp = tex1D<float>(t_img, addr);
 						if (temp > max) max = temp;
 				}
 			}
@@ -193,7 +224,19 @@ float *dilate_CUDA(int max_gicov_m, int max_gicov_n, int strel_m, int strel_n) {
 	cudaMalloc( (void**) &device_img_dilated, max_gicov_mem_size);
 	
 	// Bind the input matrix of GICOV values to a texture reference
-	cudaBindTexture(0, t_img, device_gicov, max_gicov_mem_size);
+	cudaChannelFormatDesc chDesc = cudaCreateChannelDesc<float>();
+    cudaResourceDesc resDesc = {};
+    resDesc.resType = cudaResourceTypeLinear;
+	resDesc.res.linear.devPtr = device_gicov;;
+	resDesc.res.linear.desc = chDesc;
+	resDesc.res.linear.sizeInBytes = max_gicov_mem_size;
+	cudaTextureDesc texDesc = {};
+    memset(&texDesc, 0, sizeof(texDesc));
+    texDesc.filterMode = cudaFilterModePoint;
+    texDesc.readMode = cudaReadModeElementType;
+    texDesc.normalizedCoords = 0;
+	cudaCreateTextureObject(&t_img, &resDesc, &texDesc, NULL);
+	//cudaBindTexture(0, t_img, device_gicov, max_gicov_mem_size);
     
 	// Setup execution parameters
 	int num_threads = max_gicov_m * max_gicov_n;
@@ -216,7 +259,8 @@ float *dilate_CUDA(int max_gicov_m, int max_gicov_n, int strel_m, int strel_n) {
 	cudaMemcpy(host_img_dilated, device_img_dilated, max_gicov_mem_size, cudaMemcpyDeviceToHost);
 
 	// Cleanup memory
-	cudaUnbindTexture(t_img);
+	//cudaUnbindTexture(t_img);
+	cudaDestroyTextureObject(t_img);
 	cudaFree(device_gicov);
 	cudaFree(device_img_dilated);
 
