diff --git a/common/common.mk b/common/common.mk
index 4a5d800..9dc7b6b 100644
--- a/common/common.mk
+++ b/common/common.mk
@@ -86,6 +86,7 @@ CXXWARN_FLAGS := \
 	-Wcast-align \
 	-Wreturn-type \
 	-Wno-unused-function \
+	-fPIC \
 	$(SPACE)
 
 CWARN_FLAGS := $(CXXWARN_FLAGS) \
@@ -101,7 +102,7 @@ CXXFLAGS  := $(CXXWARN_FLAGS)
 CFLAGS    := $(CWARN_FLAGS)
 
 # Common flags
-COMMONFLAGS += $(INCLUDES) -DUNIX
+COMMONFLAGS += $(INCLUDES) -DUNIX -fPIC
 
 # Debug/release configuration
 ifeq ($(dbg),1)
