diff --git a/src/blt/tests/smoke/blt_cuda_runtime_smoke.cpp b/src/blt/tests/smoke/blt_cuda_runtime_smoke.cpp
index ab505ca2249f..e188c7500da4 100644
--- a/src/blt/tests/smoke/blt_cuda_runtime_smoke.cpp
+++ b/src/blt/tests/smoke/blt_cuda_runtime_smoke.cpp
@@ -30,12 +30,16 @@ int main()
     cudaGetDeviceProperties(&prop, i);
     printf("Device Number: %d\n", i);
     printf("  Device name: %s\n", prop.name);
+#if defined(CUDART_VERSION) && CUDART_VERSION < 13000
     printf("  Memory Clock Rate (KHz): %d\n",
            prop.memoryClockRate);
+#endif
     printf("  Memory Bus Width (bits): %d\n",
            prop.memoryBusWidth);
+#if defined(CUDART_VERSION) && CUDART_VERSION < 13000
     printf("  Peak Memory Bandwidth (GB/s): %f\n\n",
            2.0*prop.memoryClockRate*(prop.memoryBusWidth/8)/1.0e6);
+#endif
   }
 
   return 0;
