diff --git a/Client/App/App.pro b/Client/App/App.pro
index 3f9c399..03a2aab 100644
--- a/Client/App/App.pro
+++ b/Client/App/App.pro
@@ -187,6 +187,9 @@ LIBS += -L../ExtendedQT -lExtendedQT
 
 LIBS += $$KVS_DIR/lib/libkvsSupportQt.a
 LIBS += $$KVS_DIR/lib/libkvsCore.a
+    equals( KVS_ENABLE_OPENGL, "1" ) {
+        LIBS += -lGL
+    }
     equals( KVS_ENABLE_GLU, "1" ) {
         LIBS += -lGLU
     }
