--- a/CS_server/main.cpp        2025-06-23 10:24:18.209279549 +0900
+++ b/CS_server/main.cpp        2025-06-23 10:26:27.981151153 +0900
@@ -1275,8 +1275,8 @@

                     for ( int tf = 0; tf < cnt; tf++ )
                     {
-                        tmp_max[tf] = 0;
-                        tmp_min[tf] = 0;
+                       tmp_max[tf] = FLT_MIN;
+                        tmp_min[tf] = FLT_MAX;
                     }

                     for ( int tf = 0; tf < c_bins_size; tf++ )
