diff --git a/singularity-eos/CMakeLists.txt b/singularity-eos/CMakeLists.txt
--- a/singularity-eos/CMakeLists.txt
+++ b/singularity-eos/CMakeLists.txt
@@ -28,6 +28,8 @@ register_headers(
     base/spiner_table_utils.hpp
     base/constants.hpp
     base/eos_error.hpp
+    base/eos_concepts.hpp
+    base/finite_diff.hpp
     eos/default_variant.hpp
     eos/eos_variant.hpp
     eos/eos_stellar_collapse.hpp
@@ -39,6 +41,7 @@ register_headers(
     eos/eos_spiner_rho_sie.hpp
     eos/eos_spiner_rho_temp.hpp
     eos/eos_spiner_common.hpp
+    eos/eos_spiner_construction.hpp
     eos/eos_spiner_sie_transforms.hpp
     eos/eos_davis.hpp
     eos/eos_gruneisen.hpp
@@ -72,6 +75,7 @@ if (SINGULARITY_BUILD_CLOSURE)
     closure/kinetic_phasetransition_models.hpp
     closure/kinetic_phasetransition_utils.hpp
     closure/kinetic_phasetransition_methods.hpp
+    closure/multiphase_material.hpp
   )
   if (SINGULARITY_BUILD_FORTRAN_BACKEND OR SINGULARITY_BUILD_TESTS)
     # while these are C++ files, they
