--- a/libgeotiff/CMakeLists.txt
+++ b/libgeotiff/CMakeLists.txt
@@ -338,11 +338,12 @@ if(TARGET ZLIB::zlib)
     string(APPEND CONFIG_PRIVATE_DEPENDENCIES "  find_dependency(ZLIB)\n")
 endif()

-target_link_libraries(${GEOTIFF_LIBRARY_TARGET} PRIVATE
-                      ${PROJ_LIBRARIES}
-                      ${TIFF_LIBRARIES}
-                      ${ZLIB_LIBRARIES})
+target_link_libraries(${GEOTIFF_LIBRARY_TARGET} PUBLIC
+                      ${PROJ_LIBRARIES})
+target_link_libraries(${GEOTIFF_LIBRARY_TARGET} PRIVATE
+                      ${TIFF_LIBRARIES}
+                      ${ZLIB_LIBRARIES})

 target_include_directories( ${GEOTIFF_LIBRARY_TARGET} PUBLIC
